// @modelcontextprotocol/client/dist/index.mjs
import { $ as INVALID_PARAMS, A as CreateMessageRequestSchema, B as JSONRPCMessageSchema, C as isJSONRPCNotification, D as parseJSONRPCMessage, E as isTaskAugmentedRequestParams, F as ElicitResultSchema, G as ListToolsResultSchema, H as ListPromptsResultSchema, I as EmptyResultSchema, J as ProtocolError, K as ReadResourceResultSchema, L as GetPromptResultSchema, M as CreateMessageResultWithToolsSchema, N as CreateTaskResultSchema, O as CallToolResultSchema, P as ElicitRequestSchema, Q as INTERNAL_ERROR, R as GetTaskPayloadResultSchema, S as isJSONRPCErrorResponse, T as isJSONRPCResultResponse, U as ListResourceTemplatesResultSchema, V as ListChangedOptionsBaseSchema, W as ListResourcesResultSchema, X as ProtocolErrorCode, Y as UrlElicitationRequiredError, Z as DEFAULT_NEGOTIATED_PROTOCOL_VERSION, _ as parseSchema, _t as SdkError, a as createFetchWithInit, at as RELATED_TASK_META_KEY, b as isInitializeRequest, bt as OAuthErrorCode, c as deserializeMessage, ct as checkResourceAllowed, d as toArrayAsync, dt as OAuthClientInformationFullSchema, et as INVALID_REQUEST, f as DEFAULT_REQUEST_TIMEOUT_MSEC, ft as OAuthErrorResponseSchema, g as isTerminal, gt as OpenIdProviderDiscoveryMetadataSchema, h as extractTaskManagerOptions, ht as OAuthTokensSchema, i as UriTemplate, it as PARSE_ERROR, j as CreateMessageResultSchema, k as CompleteResultSchema, l as serializeMessage, lt as resourceUrlFromServerUrl, m as mergeCapabilities, mt as OAuthProtectedResourceMetadataSchema, n as CfWorkerJsonSchemaValidator, nt as LATEST_PROTOCOL_VERSION, o as normalizeHeaders, ot as SUPPORTED_PROTOCOL_VERSIONS, p as Protocol, pt as OAuthMetadataSchema, q as getResultSchema, r as AjvJsonSchemaValidator, rt as METHOD_NOT_FOUND, s as ReadBuffer, st as getDisplayName, t as fromJsonSchema$1, tt as JSONRPC_VERSION, u as takeResult, ut as IdJagTokenExchangeResponseSchema, v as assertCompleteRequestPrompt, vt as SdkErrorCode, w as isJSONRPCRequest, x as isInitializedNotification, y as assertCompleteRequestResourceTemplate, yt as OAuthError, z as InitializeResultSchema } from "./src-RIHfgPRm.mjs";
import { CORS_IS_POSSIBLE, DefaultJsonSchemaValidator } from "@modelcontextprotocol/client/_shims";
import pkceChallenge from "pkce-challenge";
import { EventSource } from "eventsource";
import process from "node:process";
import { PassThrough } from "node:stream";
import spawn from "cross-spawn";
import { EventSourceParserStream } from "eventsource-parser/stream";

//#region ../core/src/experimental/tasks/helpers.ts
/**
* Experimental task capability assertion helpers.
* WARNING: These APIs are experimental and may change without notice.
*
* @experimental
*/
/**
* Asserts that task creation is supported for `tools/call`.
* Used to implement the `assertTaskCapability` or `assertTaskHandlerCapability` abstract methods on Protocol.
*
* @param requests - The task requests capability object
* @param method - The method being checked
* @param entityName - `'Server'` or `'Client'` for error messages
* @throws {@linkcode SdkError} with {@linkcode SdkErrorCode.CapabilityNotSupported} if the capability is not supported
*
* @experimental
*/
function assertToolsCallTaskCapability(requests, method, entityName) {
	if (!requests) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `${entityName} does not support task creation (required for ${method})`);
	switch (method) {
		case "tools/call":
			if (!requests.tools?.call) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `${entityName} does not support task creation for tools/call (required for ${method})`);
			break;
		default: break;
	}
}
/**
* Asserts that task creation is supported for `sampling/createMessage` or `elicitation/create`.
* Used to implement the `assertTaskCapability` or `assertTaskHandlerCapability` abstract methods on Protocol.
*
* @param requests - The task requests capability object
* @param method - The method being checked
* @param entityName - `'Server'` or `'Client'` for error messages
* @throws {@linkcode SdkError} with {@linkcode SdkErrorCode.CapabilityNotSupported} if the capability is not supported
*
* @experimental
*/
function assertClientRequestTaskCapability(requests, method, entityName) {
	if (!requests) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `${entityName} does not support task creation (required for ${method})`);
	switch (method) {
		case "sampling/createMessage":
			if (!requests.sampling?.createMessage) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `${entityName} does not support task creation for sampling/createMessage (required for ${method})`);
			break;
		case "elicitation/create":
			if (!requests.elicitation?.create) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `${entityName} does not support task creation for elicitation/create (required for ${method})`);
			break;
		default: break;
	}
}

//#endregion
//#region ../core/src/experimental/tasks/stores/inMemory.ts
/**
* In-memory {@linkcode TaskStore} implementation for development and testing.
* For production, use a database or distributed cache.
* @experimental
*/
var InMemoryTaskStore = class {
	tasks = /* @__PURE__ */ new Map();
	cleanupTimers = /* @__PURE__ */ new Map();
	/**
	* Generates a unique task ID using Web Crypto API.
	*/
	generateTaskId() {
		return crypto.randomUUID().replaceAll("-", "");
	}
	/** {@inheritDoc TaskStore.createTask} */
	async createTask(taskParams, requestId, request, sessionId) {
		const taskId = this.generateTaskId();
		if (this.tasks.has(taskId)) throw new Error(`Task with ID ${taskId} already exists`);
		const actualTtl = taskParams.ttl ?? null;
		const createdAt = (/* @__PURE__ */ new Date()).toISOString();
		const task = {
			taskId,
			status: "working",
			ttl: actualTtl,
			createdAt,
			lastUpdatedAt: createdAt,
			pollInterval: taskParams.pollInterval ?? 1e3
		};
		this.tasks.set(taskId, {
			task,
			request,
			requestId,
			sessionId
		});
		if (actualTtl) {
			const timer = setTimeout(() => {
				this.tasks.delete(taskId);
				this.cleanupTimers.delete(taskId);
			}, actualTtl);
			this.cleanupTimers.set(taskId, timer);
		}
		return task;
	}
	/**
	* Retrieves a stored task, enforcing session ownership when a sessionId is provided.
	* Returns undefined if the task does not exist or belongs to a different session.
	*/
	getStoredTask(taskId, sessionId) {
		const stored = this.tasks.get(taskId);
		if (!stored) return;
		if (sessionId !== void 0 && stored.sessionId !== void 0 && stored.sessionId !== sessionId) return;
		return stored;
	}
	async getTask(taskId, sessionId) {
		const stored = this.getStoredTask(taskId, sessionId);
		return stored ? { ...stored.task } : null;
	}
	/** {@inheritDoc TaskStore.storeTaskResult} */
	async storeTaskResult(taskId, status, result, sessionId) {
		const stored = this.getStoredTask(taskId, sessionId);
		if (!stored) throw new Error(`Task with ID ${taskId} not found`);
		if (isTerminal(stored.task.status)) throw new Error(`Cannot store result for task ${taskId} in terminal status '${stored.task.status}'. Task results can only be stored once.`);
		stored.result = result;
		stored.task.status = status;
		stored.task.lastUpdatedAt = (/* @__PURE__ */ new Date()).toISOString();
		if (stored.task.ttl) {
			const existingTimer = this.cleanupTimers.get(taskId);
			if (existingTimer) clearTimeout(existingTimer);
			const timer = setTimeout(() => {
				this.tasks.delete(taskId);
				this.cleanupTimers.delete(taskId);
			}, stored.task.ttl);
			this.cleanupTimers.set(taskId, timer);
		}
	}
	/** {@inheritDoc TaskStore.getTaskResult} */
	async getTaskResult(taskId, sessionId) {
		const stored = this.getStoredTask(taskId, sessionId);
		if (!stored) throw new Error(`Task with ID ${taskId} not found`);
		if (!stored.result) throw new Error(`Task ${taskId} has no result stored`);
		return stored.result;
	}
	/** {@inheritDoc TaskStore.updateTaskStatus} */
	async updateTaskStatus(taskId, status, statusMessage, sessionId) {
		const stored = this.getStoredTask(taskId, sessionId);
		if (!stored) throw new Error(`Task with ID ${taskId} not found`);
		if (isTerminal(stored.task.status)) throw new Error(`Cannot update task ${taskId} from terminal status '${stored.task.status}' to '${status}'. Terminal states (completed, failed, cancelled) cannot transition to other states.`);
		stored.task.status = status;
		if (statusMessage) stored.task.statusMessage = statusMessage;
		stored.task.lastUpdatedAt = (/* @__PURE__ */ new Date()).toISOString();
		if (isTerminal(status) && stored.task.ttl) {
			const existingTimer = this.cleanupTimers.get(taskId);
			if (existingTimer) clearTimeout(existingTimer);
			const timer = setTimeout(() => {
				this.tasks.delete(taskId);
				this.cleanupTimers.delete(taskId);
			}, stored.task.ttl);
			this.cleanupTimers.set(taskId, timer);
		}
	}
	/** {@inheritDoc TaskStore.listTasks} */
	async listTasks(cursor, sessionId) {
		const PAGE_SIZE = 10;
		const filteredTaskIds = [...this.tasks.entries()].filter(([, stored]) => {
			if (sessionId === void 0 || stored.sessionId === void 0) return true;
			return stored.sessionId === sessionId;
		}).map(([taskId]) => taskId);
		let startIndex = 0;
		if (cursor) {
			const cursorIndex = filteredTaskIds.indexOf(cursor);
			if (cursorIndex === -1) throw new Error(`Invalid cursor: ${cursor}`);
			else startIndex = cursorIndex + 1;
		}
		const pageTaskIds = filteredTaskIds.slice(startIndex, startIndex + PAGE_SIZE);
		return {
			tasks: pageTaskIds.map((taskId) => {
				return { ...this.tasks.get(taskId).task };
			}),
			nextCursor: startIndex + PAGE_SIZE < filteredTaskIds.length ? pageTaskIds.at(-1) : void 0
		};
	}
	/**
	* Cleanup all timers (useful for testing or graceful shutdown)
	*/
	cleanup() {
		for (const timer of this.cleanupTimers.values()) clearTimeout(timer);
		this.cleanupTimers.clear();
		this.tasks.clear();
	}
	/**
	* Get all tasks (useful for debugging)
	*/
	getAllTasks() {
		return [...this.tasks.values()].map((stored) => ({ ...stored.task }));
	}
};
/**
* In-memory {@linkcode TaskMessageQueue} implementation for development and testing.
* For production, use Redis or another distributed queue.
* @experimental
*/
var InMemoryTaskMessageQueue = class {
	queues = /* @__PURE__ */ new Map();
	/**
	* Generates a queue key from taskId.
	* SessionId is intentionally ignored because taskIds are globally unique
	* and tasks need to be accessible across HTTP requests/sessions.
	*/
	getQueueKey(taskId, _sessionId) {
		return taskId;
	}
	/**
	* Gets or creates a queue for the given task and session.
	*/
	getQueue(taskId, sessionId) {
		const key = this.getQueueKey(taskId, sessionId);
		let queue = this.queues.get(key);
		if (!queue) {
			queue = [];
			this.queues.set(key, queue);
		}
		return queue;
	}
	/**
	* Adds a message to the end of the queue for a specific task.
	* Atomically checks queue size and throws if maxSize would be exceeded.
	* @param taskId The task identifier
	* @param message The message to enqueue
	* @param sessionId Optional session ID for binding the operation to a specific session
	* @param maxSize Optional maximum queue size - if specified and queue is full, throws an error
	* @throws Error if maxSize is specified and would be exceeded
	*/
	async enqueue(taskId, message, sessionId, maxSize) {
		const queue = this.getQueue(taskId, sessionId);
		if (maxSize !== void 0 && queue.length >= maxSize) throw new Error(`Task message queue overflow: queue size (${queue.length}) exceeds maximum (${maxSize})`);
		queue.push(message);
	}
	/**
	* Removes and returns the first message from the queue for a specific task.
	* @param taskId The task identifier
	* @param sessionId Optional session ID for binding the query to a specific session
	* @returns The first message, or `undefined` if the queue is empty
	*/
	async dequeue(taskId, sessionId) {
		return this.getQueue(taskId, sessionId).shift();
	}
	/**
	* Removes and returns all messages from the queue for a specific task.
	* @param taskId The task identifier
	* @param sessionId Optional session ID for binding the query to a specific session
	* @returns Array of all messages that were in the queue
	*/
	async dequeueAll(taskId, sessionId) {
		const key = this.getQueueKey(taskId, sessionId);
		const queue = this.queues.get(key) ?? [];
		this.queues.delete(key);
		return queue;
	}
};

//#endregion
//#region src/client/auth.ts
/**
* Type guard distinguishing `OAuthClientProvider` from a minimal `AuthProvider`.
* Transports use this at construction time to classify the `authProvider` option.
*
* Checks for `tokens()` + `clientInformation()` — two required `OAuthClientProvider`
* methods that a minimal `AuthProvider` `{ token: ... }` would never have.
*/
function isOAuthClientProvider(provider) {
	if (provider == null) return false;
	const p = provider;
	return typeof p.tokens === "function" && typeof p.clientInformation === "function";
}
/**
* Standard `onUnauthorized` behavior for OAuth providers: extracts
* `WWW-Authenticate` parameters from the 401 response and runs {@linkcode auth}.
* Used by {@linkcode adaptOAuthProvider} to bridge `OAuthClientProvider` to `AuthProvider`.
*/
async function handleOAuthUnauthorized(provider, ctx) {
	const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(ctx.response);
	if (await auth(provider, {
		serverUrl: ctx.serverUrl,
		resourceMetadataUrl,
		scope,
		fetchFn: ctx.fetchFn
	}) !== "AUTHORIZED") throw new UnauthorizedError();
}
/**
* Adapts an `OAuthClientProvider` to the minimal `AuthProvider` interface that
* transports consume. Called once at transport construction — the transport stores
* the adapted provider for `_commonHeaders()` and 401 handling, while keeping the
* original `OAuthClientProvider` for OAuth-specific paths (`finishAuth()`, 403 upscoping).
*/
function adaptOAuthProvider(provider) {
	return {
		token: async () => {
			return (await provider.tokens())?.access_token;
		},
		onUnauthorized: async (ctx) => handleOAuthUnauthorized(provider, ctx)
	};
}
var UnauthorizedError = class extends Error {
	constructor(message) {
		super(message ?? "Unauthorized");
	}
};
function isClientAuthMethod(method) {
	return [
		"client_secret_basic",
		"client_secret_post",
		"none"
	].includes(method);
}
const AUTHORIZATION_CODE_RESPONSE_TYPE = "code";
const AUTHORIZATION_CODE_CHALLENGE_METHOD = "S256";
/**
* Determines the best client authentication method to use based on server support and client configuration.
*
* Priority order (highest to lowest):
* 1. `client_secret_basic` (if client secret is available)
* 2. `client_secret_post` (if client secret is available)
* 3. `none` (for public clients)
*
* @param clientInformation - OAuth client information containing credentials
* @param supportedMethods - Authentication methods supported by the authorization server
* @returns The selected authentication method
*/
function selectClientAuthMethod(clientInformation, supportedMethods) {
	const hasClientSecret = clientInformation.client_secret !== void 0;
	if ("token_endpoint_auth_method" in clientInformation && clientInformation.token_endpoint_auth_method && isClientAuthMethod(clientInformation.token_endpoint_auth_method) && (supportedMethods.length === 0 || supportedMethods.includes(clientInformation.token_endpoint_auth_method))) return clientInformation.token_endpoint_auth_method;
	if (supportedMethods.length === 0) return hasClientSecret ? "client_secret_basic" : "none";
	if (hasClientSecret && supportedMethods.includes("client_secret_basic")) return "client_secret_basic";
	if (hasClientSecret && supportedMethods.includes("client_secret_post")) return "client_secret_post";
	if (supportedMethods.includes("none")) return "none";
	return hasClientSecret ? "client_secret_post" : "none";
}
/**
* Applies client authentication to the request based on the specified method.
*
* Implements OAuth 2.1 client authentication methods:
* - `client_secret_basic`: HTTP Basic authentication (RFC 6749 Section 2.3.1)
* - `client_secret_post`: Credentials in request body (RFC 6749 Section 2.3.1)
* - `none`: Public client authentication (RFC 6749 Section 2.1)
*
* @param method - The authentication method to use
* @param clientInformation - OAuth client information containing credentials
* @param headers - HTTP headers object to modify
* @param params - URL search parameters to modify
* @throws {Error} When required credentials are missing
*/
function applyClientAuthentication(method, clientInformation, headers, params) {
	const { client_id, client_secret } = clientInformation;
	switch (method) {
		case "client_secret_basic":
			applyBasicAuth(client_id, client_secret, headers);
			return;
		case "client_secret_post":
			applyPostAuth(client_id, client_secret, params);
			return;
		case "none":
			applyPublicAuth(client_id, params);
			return;
		default: throw new Error(`Unsupported client authentication method: ${method}`);
	}
}
/**
* Applies HTTP Basic authentication (RFC 6749 Section 2.3.1)
*/
function applyBasicAuth(clientId, clientSecret, headers) {
	if (!clientSecret) throw new Error("client_secret_basic authentication requires a client_secret");
	const credentials = btoa(`${clientId}:${clientSecret}`);
	headers.set("Authorization", `Basic ${credentials}`);
}
/**
* Applies POST body authentication (RFC 6749 Section 2.3.1)
*/
function applyPostAuth(clientId, clientSecret, params) {
	params.set("client_id", clientId);
	if (clientSecret) params.set("client_secret", clientSecret);
}
/**
* Applies public client authentication (RFC 6749 Section 2.1)
*/
function applyPublicAuth(clientId, params) {
	params.set("client_id", clientId);
}
/**
* Parses an OAuth error response from a string or Response object.
*
* If the input is a standard OAuth2.0 error response, it will be parsed according to the spec
* and an {@linkcode OAuthError} will be returned with the appropriate error code.
* If parsing fails, it falls back to a generic {@linkcode OAuthErrorCode.ServerError | ServerError} that includes
* the response status (if available) and original content.
*
* @param input - A Response object or string containing the error response
* @returns A Promise that resolves to an {@linkcode OAuthError} instance
*/
async function parseErrorResponse(input) {
	const statusCode = input instanceof Response ? input.status : void 0;
	const body = input instanceof Response ? await input.text() : input;
	try {
		const result = OAuthErrorResponseSchema.parse(JSON.parse(body));
		return OAuthError.fromResponse(result);
	} catch (error) {
		const errorMessage = `${statusCode ? `HTTP ${statusCode}: ` : ""}Invalid OAuth error response: ${error}. Raw body: ${body}`;
		return new OAuthError(OAuthErrorCode.ServerError, errorMessage);
	}
}
/**
* Orchestrates the full auth flow with a server.
*
* This can be used as a single entry point for all authorization functionality,
* instead of linking together the other lower-level functions in this module.
*/
async function auth(provider, options) {
	try {
		return await authInternal(provider, options);
	} catch (error) {
		if (error instanceof OAuthError) {
			if (error.code === OAuthErrorCode.InvalidClient || error.code === OAuthErrorCode.UnauthorizedClient) {
				await provider.invalidateCredentials?.("all");
				return await authInternal(provider, options);
			} else if (error.code === OAuthErrorCode.InvalidGrant) {
				await provider.invalidateCredentials?.("tokens");
				return await authInternal(provider, options);
			}
		}
		throw error;
	}
}
/**
* Selects scopes per the MCP spec and augment for refresh token support.
*/
function determineScope(options) {
	const { requestedScope, resourceMetadata, authServerMetadata, clientMetadata } = options;
	let effectiveScope = requestedScope || resourceMetadata?.scopes_supported?.join(" ") || clientMetadata.scope;
	if (effectiveScope && authServerMetadata?.scopes_supported?.includes("offline_access") && !effectiveScope.split(" ").includes("offline_access") && clientMetadata.grant_types?.includes("refresh_token")) effectiveScope = `${effectiveScope} offline_access`;
	return effectiveScope;
}
async function authInternal(provider, { serverUrl, authorizationCode, scope, resourceMetadataUrl, fetchFn }) {
	const cachedState = await provider.discoveryState?.();
	let resourceMetadata;
	let authorizationServerUrl;
	let metadata;
	let effectiveResourceMetadataUrl = resourceMetadataUrl;
	if (!effectiveResourceMetadataUrl && cachedState?.resourceMetadataUrl) effectiveResourceMetadataUrl = new URL(cachedState.resourceMetadataUrl);
	if (cachedState?.authorizationServerUrl) {
		authorizationServerUrl = cachedState.authorizationServerUrl;
		resourceMetadata = cachedState.resourceMetadata;
		metadata = cachedState.authorizationServerMetadata ?? await discoverAuthorizationServerMetadata(authorizationServerUrl, { fetchFn });
		if (!resourceMetadata) try {
			resourceMetadata = await discoverOAuthProtectedResourceMetadata(serverUrl, { resourceMetadataUrl: effectiveResourceMetadataUrl }, fetchFn);
		} catch (error) {
			if (error instanceof TypeError) throw error;
		}
		if (metadata !== cachedState.authorizationServerMetadata || resourceMetadata !== cachedState.resourceMetadata) await provider.saveDiscoveryState?.({
			authorizationServerUrl: String(authorizationServerUrl),
			resourceMetadataUrl: effectiveResourceMetadataUrl?.toString(),
			resourceMetadata,
			authorizationServerMetadata: metadata
		});
	} else {
		const serverInfo = await discoverOAuthServerInfo(serverUrl, {
			resourceMetadataUrl: effectiveResourceMetadataUrl,
			fetchFn
		});
		authorizationServerUrl = serverInfo.authorizationServerUrl;
		metadata = serverInfo.authorizationServerMetadata;
		resourceMetadata = serverInfo.resourceMetadata;
		await provider.saveDiscoveryState?.({
			authorizationServerUrl: String(authorizationServerUrl),
			resourceMetadataUrl: effectiveResourceMetadataUrl?.toString(),
			resourceMetadata,
			authorizationServerMetadata: metadata
		});
	}
	await provider.saveAuthorizationServerUrl?.(String(authorizationServerUrl));
	const resource = await selectResourceURL(serverUrl, provider, resourceMetadata);
	if (resource) await provider.saveResourceUrl?.(String(resource));
	const resolvedScope = determineScope({
		requestedScope: scope,
		resourceMetadata,
		authServerMetadata: metadata,
		clientMetadata: provider.clientMetadata
	});
	let clientInformation = await Promise.resolve(provider.clientInformation());
	if (!clientInformation) {
		if (authorizationCode !== void 0) throw new Error("Existing OAuth client information is required when exchanging an authorization code");
		const supportsUrlBasedClientId = metadata?.client_id_metadata_document_supported === true;
		const clientMetadataUrl = provider.clientMetadataUrl;
		if (clientMetadataUrl && !isHttpsUrl(clientMetadataUrl)) throw new OAuthError(OAuthErrorCode.InvalidClientMetadata, `clientMetadataUrl must be a valid HTTPS URL with a non-root pathname, got: ${clientMetadataUrl}`);
		if (supportsUrlBasedClientId && clientMetadataUrl) {
			clientInformation = { client_id: clientMetadataUrl };
			await provider.saveClientInformation?.(clientInformation);
		} else {
			if (!provider.saveClientInformation) throw new Error("OAuth client information must be saveable for dynamic registration");
			const fullInformation = await registerClient(authorizationServerUrl, {
				metadata,
				clientMetadata: provider.clientMetadata,
				scope: resolvedScope,
				fetchFn
			});
			await provider.saveClientInformation(fullInformation);
			clientInformation = fullInformation;
		}
	}
	const nonInteractiveFlow = !provider.redirectUrl;
	if (authorizationCode !== void 0 || nonInteractiveFlow) {
		const tokens$1 = await fetchToken(provider, authorizationServerUrl, {
			metadata,
			resource,
			authorizationCode,
			scope: resolvedScope,
			fetchFn
		});
		await provider.saveTokens(tokens$1);
		return "AUTHORIZED";
	}
	const tokens = await provider.tokens();
	if (tokens?.refresh_token) try {
		const newTokens = await refreshAuthorization(authorizationServerUrl, {
			metadata,
			clientInformation,
			refreshToken: tokens.refresh_token,
			resource,
			addClientAuthentication: provider.addClientAuthentication,
			fetchFn
		});
		await provider.saveTokens(newTokens);
		return "AUTHORIZED";
	} catch (error) {
		if (!(error instanceof OAuthError) || error.code === OAuthErrorCode.ServerError) {} else throw error;
	}
	const state = provider.state ? await provider.state() : void 0;
	const { authorizationUrl, codeVerifier } = await startAuthorization(authorizationServerUrl, {
		metadata,
		clientInformation,
		state,
		redirectUrl: provider.redirectUrl,
		scope: resolvedScope,
		resource
	});
	await provider.saveCodeVerifier(codeVerifier);
	await provider.redirectToAuthorization(authorizationUrl);
	return "REDIRECT";
}
/**
* SEP-991: URL-based Client IDs
* Validate that the `client_id` is a valid URL with `https` scheme
*/
function isHttpsUrl(value) {
	if (!value) return false;
	try {
		const url = new URL(value);
		return url.protocol === "https:" && url.pathname !== "/";
	} catch {
		return false;
	}
}
async function selectResourceURL(serverUrl, provider, resourceMetadata) {
	const defaultResource = resourceUrlFromServerUrl(serverUrl);
	if (provider.validateResourceURL) return await provider.validateResourceURL(defaultResource, resourceMetadata?.resource);
	if (!resourceMetadata) return;
	if (!checkResourceAllowed({
		requestedResource: defaultResource,
		configuredResource: resourceMetadata.resource
	})) throw new Error(`Protected resource ${resourceMetadata.resource} does not match expected ${defaultResource} (or origin)`);
	return new URL(resourceMetadata.resource);
}
/**
* Extract `resource_metadata`, `scope`, and `error` from `WWW-Authenticate` header.
*/
function extractWWWAuthenticateParams(res) {
	const authenticateHeader = res.headers.get("WWW-Authenticate");
	if (!authenticateHeader) return {};
	const [type, scheme] = authenticateHeader.split(" ");
	if (type?.toLowerCase() !== "bearer" || !scheme) return {};
	const resourceMetadataMatch = extractFieldFromWwwAuth(res, "resource_metadata") || void 0;
	let resourceMetadataUrl;
	if (resourceMetadataMatch) try {
		resourceMetadataUrl = new URL(resourceMetadataMatch);
	} catch {}
	const scope = extractFieldFromWwwAuth(res, "scope") || void 0;
	const error = extractFieldFromWwwAuth(res, "error") || void 0;
	return {
		resourceMetadataUrl,
		scope,
		error
	};
}
/**
* Extracts a specific field's value from the `WWW-Authenticate` header string.
*
* @param response The HTTP response object containing the headers.
* @param fieldName The name of the field to extract (e.g., `"realm"`, `"nonce"`).
* @returns The field value
*/
function extractFieldFromWwwAuth(response, fieldName) {
	const wwwAuthHeader = response.headers.get("WWW-Authenticate");
	if (!wwwAuthHeader) return null;
	const pattern = new RegExp(String.raw`${fieldName}=(?:"([^"]+)"|([^\s,]+))`);
	const match = wwwAuthHeader.match(pattern);
	if (match) {
		const result = match[1] || match[2];
		if (result) return result;
	}
	return null;
}
/**
* Extract `resource_metadata` from response header.
* @deprecated Use {@linkcode extractWWWAuthenticateParams} instead.
*/
function extractResourceMetadataUrl(res) {
	const authenticateHeader = res.headers.get("WWW-Authenticate");
	if (!authenticateHeader) return;
	const [type, scheme] = authenticateHeader.split(" ");
	if (type?.toLowerCase() !== "bearer" || !scheme) return;
	const match = /resource_metadata="([^"]*)"/.exec(authenticateHeader);
	if (!match || !match[1]) return;
	try {
		return new URL(match[1]);
	} catch {
		return;
	}
}
/**
* Looks up {@link https://datatracker.ietf.org/doc/html/rfc9728 | RFC 9728}
* OAuth 2.0 Protected Resource Metadata.
*
* If the server returns a 404 for the well-known endpoint, this function will
* return `undefined`. Any other errors will be thrown as exceptions.
*/
async function discoverOAuthProtectedResourceMetadata(serverUrl, opts, fetchFn = fetch) {
	const response = await discoverMetadataWithFallback(serverUrl, "oauth-protected-resource", fetchFn, {
		protocolVersion: opts?.protocolVersion,
		metadataUrl: opts?.resourceMetadataUrl
	});
	if (!response || response.status === 404) {
		await response?.text?.().catch(() => {});
		throw new Error(`Resource server does not implement OAuth 2.0 Protected Resource Metadata.`);
	}
	if (!response.ok) {
		await response.text?.().catch(() => {});
		throw new Error(`HTTP ${response.status} trying to load well-known OAuth protected resource metadata.`);
	}
	return OAuthProtectedResourceMetadataSchema.parse(await response.json());
}
/**
* Fetch with a retry heuristic for CORS errors caused by custom headers.
*
* In browsers, adding a custom header (e.g. `MCP-Protocol-Version`) triggers a CORS preflight.
* If the server doesn't allow that header, the browser throws a `TypeError` before any response
* is received. Retrying without custom headers often succeeds because the request becomes
* "simple" (no preflight). If the server sends no CORS headers at all, the retry also fails
* with `TypeError` and we return `undefined` so callers can fall through to an alternate URL.
*
* However, `fetch()` also throws `TypeError` for non-CORS failures (DNS resolution, connection
* refused, invalid URL). Swallowing those and returning `undefined` masks real errors and can
* cause callers to silently fall through to a different discovery URL. CORS is a browser-only
* concept, so in non-browser runtimes (Node.js, Workers) a `TypeError` from `fetch` is never a
* CORS error — there we propagate the error instead of swallowing it.
*
* In browsers, we cannot reliably distinguish CORS `TypeError` from network `TypeError` from the
* error object alone, so the swallow-and-fallthrough heuristic is preserved there.
*/
async function fetchWithCorsRetry(url, headers, fetchFn = fetch) {
	try {
		return await fetchFn(url, { headers });
	} catch (error) {
		if (!(error instanceof TypeError) || !CORS_IS_POSSIBLE) throw error;
		if (headers) try {
			return await fetchFn(url, {});
		} catch (retryError) {
			if (!(retryError instanceof TypeError)) throw retryError;
			return;
		}
		return;
	}
}
/**
* Constructs the well-known path for auth-related metadata discovery
*/
function buildWellKnownPath(wellKnownPrefix, pathname = "", options = {}) {
	if (pathname.endsWith("/")) pathname = pathname.slice(0, -1);
	return options.prependPathname ? `${pathname}/.well-known/${wellKnownPrefix}` : `/.well-known/${wellKnownPrefix}${pathname}`;
}
/**
* Tries to discover OAuth metadata at a specific URL
*/
async function tryMetadataDiscovery(url, protocolVersion, fetchFn = fetch) {
	return await fetchWithCorsRetry(url, { "MCP-Protocol-Version": protocolVersion }, fetchFn);
}
/**
* Determines if fallback to root discovery should be attempted
*/
function shouldAttemptFallback(response, pathname) {
	if (!response) return true;
	if (pathname === "/") return false;
	return response.status >= 400 && response.status < 500 || response.status === 502;
}
/**
* Generic function for discovering OAuth metadata with fallback support
*/
async function discoverMetadataWithFallback(serverUrl, wellKnownType, fetchFn, opts) {
	const issuer = new URL(serverUrl);
	const protocolVersion = opts?.protocolVersion ?? LATEST_PROTOCOL_VERSION;
	let url;
	if (opts?.metadataUrl) url = new URL(opts.metadataUrl);
	else {
		const wellKnownPath = buildWellKnownPath(wellKnownType, issuer.pathname);
		url = new URL(wellKnownPath, opts?.metadataServerUrl ?? issuer);
		url.search = issuer.search;
	}
	let response = await tryMetadataDiscovery(url, protocolVersion, fetchFn);
	if (!opts?.metadataUrl && shouldAttemptFallback(response, issuer.pathname)) response = await tryMetadataDiscovery(new URL(`/.well-known/${wellKnownType}`, issuer), protocolVersion, fetchFn);
	return response;
}
/**
* Looks up RFC 8414 OAuth 2.0 Authorization Server Metadata.
*
* If the server returns a 404 for the well-known endpoint, this function will
* return `undefined`. Any other errors will be thrown as exceptions.
*
* @deprecated This function is deprecated in favor of {@linkcode discoverAuthorizationServerMetadata}.
*/
async function discoverOAuthMetadata(issuer, { authorizationServerUrl, protocolVersion } = {}, fetchFn = fetch) {
	if (typeof issuer === "string") issuer = new URL(issuer);
	if (!authorizationServerUrl) authorizationServerUrl = issuer;
	if (typeof authorizationServerUrl === "string") authorizationServerUrl = new URL(authorizationServerUrl);
	protocolVersion ??= LATEST_PROTOCOL_VERSION;
	const response = await discoverMetadataWithFallback(authorizationServerUrl, "oauth-authorization-server", fetchFn, {
		protocolVersion,
		metadataServerUrl: authorizationServerUrl
	});
	if (!response || response.status === 404) {
		await response?.text?.().catch(() => {});
		return;
	}
	if (!response.ok) {
		await response.text?.().catch(() => {});
		throw new Error(`HTTP ${response.status} trying to load well-known OAuth metadata`);
	}
	return OAuthMetadataSchema.parse(await response.json());
}
/**
* Builds a list of discovery URLs to try for authorization server metadata.
* URLs are returned in priority order:
* 1. OAuth metadata at the given URL
* 2. OIDC metadata endpoints at the given URL
*/
function buildDiscoveryUrls(authorizationServerUrl) {
	const url = typeof authorizationServerUrl === "string" ? new URL(authorizationServerUrl) : authorizationServerUrl;
	const hasPath = url.pathname !== "/";
	const urlsToTry = [];
	if (!hasPath) {
		urlsToTry.push({
			url: new URL("/.well-known/oauth-authorization-server", url.origin),
			type: "oauth"
		}, {
			url: new URL(`/.well-known/openid-configuration`, url.origin),
			type: "oidc"
		});
		return urlsToTry;
	}
	let pathname = url.pathname;
	if (pathname.endsWith("/")) pathname = pathname.slice(0, -1);
	urlsToTry.push({
		url: new URL(`/.well-known/oauth-authorization-server${pathname}`, url.origin),
		type: "oauth"
	}, {
		url: new URL(`/.well-known/openid-configuration${pathname}`, url.origin),
		type: "oidc"
	}, {
		url: new URL(`${pathname}/.well-known/openid-configuration`, url.origin),
		type: "oidc"
	});
	return urlsToTry;
}
/**
* Discovers authorization server metadata with support for
* {@link https://datatracker.ietf.org/doc/html/rfc8414 | RFC 8414} OAuth 2.0
* Authorization Server Metadata and
* {@link https://openid.net/specs/openid-connect-discovery-1_0.html | OpenID Connect Discovery 1.0}
* specifications.
*
* This function implements a fallback strategy for authorization server discovery:
* 1. Attempts RFC 8414 OAuth metadata discovery first
* 2. If OAuth discovery fails, falls back to OpenID Connect Discovery
*
* @param authorizationServerUrl - The authorization server URL obtained from the MCP Server's
*                                 protected resource metadata, or the MCP server's URL if the
*                                 metadata was not found.
* @param options - Configuration options
* @param options.fetchFn - Optional fetch function for making HTTP requests, defaults to global fetch
* @param options.protocolVersion - MCP protocol version to use, defaults to {@linkcode LATEST_PROTOCOL_VERSION}
* @returns Promise resolving to authorization server metadata, or undefined if discovery fails
*/
async function discoverAuthorizationServerMetadata(authorizationServerUrl, { fetchFn = fetch, protocolVersion = LATEST_PROTOCOL_VERSION } = {}) {
	const headers = {
		"MCP-Protocol-Version": protocolVersion,
		Accept: "application/json"
	};
	const urlsToTry = buildDiscoveryUrls(authorizationServerUrl);
	for (const { url: endpointUrl, type } of urlsToTry) {
		const response = await fetchWithCorsRetry(endpointUrl, headers, fetchFn);
		if (!response)
 /**
		* CORS error occurred - don't throw as the endpoint may not allow CORS,
		* continue trying other possible endpoints
		*/
		continue;
		if (!response.ok) {
			await response.text?.().catch(() => {});
			if (response.status >= 400 && response.status < 500 || response.status === 502) continue;
			throw new Error(`HTTP ${response.status} trying to load ${type === "oauth" ? "OAuth" : "OpenID provider"} metadata from ${endpointUrl}`);
		}
		return type === "oauth" ? OAuthMetadataSchema.parse(await response.json()) : OpenIdProviderDiscoveryMetadataSchema.parse(await response.json());
	}
}
/**
* Discovers the authorization server for an MCP server following
* {@link https://datatracker.ietf.org/doc/html/rfc9728 | RFC 9728} (OAuth 2.0 Protected
* Resource Metadata), with fallback to treating the server URL as the
* authorization server.
*
* This function combines two discovery steps into one call:
* 1. Probes `/.well-known/oauth-protected-resource` on the MCP server to find the
*    authorization server URL (RFC 9728).
* 2. Fetches authorization server metadata from that URL (RFC 8414 / OpenID Connect Discovery).
*
* Use this when you need the authorization server metadata for operations outside the
* {@linkcode auth} orchestrator, such as token refresh or token revocation.
*
* @param serverUrl - The MCP resource server URL
* @param opts - Optional configuration
* @param opts.resourceMetadataUrl - Override URL for the protected resource metadata endpoint
* @param opts.fetchFn - Custom fetch function for HTTP requests
* @returns Authorization server URL, metadata, and resource metadata (if available)
*/
async function discoverOAuthServerInfo(serverUrl, opts) {
	let resourceMetadata;
	let authorizationServerUrl;
	try {
		resourceMetadata = await discoverOAuthProtectedResourceMetadata(serverUrl, { resourceMetadataUrl: opts?.resourceMetadataUrl }, opts?.fetchFn);
		if (resourceMetadata.authorization_servers && resourceMetadata.authorization_servers.length > 0) authorizationServerUrl = resourceMetadata.authorization_servers[0];
	} catch (error) {
		if (error instanceof TypeError) throw error;
	}
	if (!authorizationServerUrl) authorizationServerUrl = String(new URL("/", serverUrl));
	const authorizationServerMetadata = await discoverAuthorizationServerMetadata(authorizationServerUrl, { fetchFn: opts?.fetchFn });
	return {
		authorizationServerUrl,
		authorizationServerMetadata,
		resourceMetadata
	};
}
/**
* Begins the authorization flow with the given server, by generating a PKCE challenge and constructing the authorization URL.
*/
async function startAuthorization(authorizationServerUrl, { metadata, clientInformation, redirectUrl, scope, state, resource }) {
	let authorizationUrl;
	if (metadata) {
		authorizationUrl = new URL(metadata.authorization_endpoint);
		if (!metadata.response_types_supported.includes(AUTHORIZATION_CODE_RESPONSE_TYPE)) throw new Error(`Incompatible auth server: does not support response type ${AUTHORIZATION_CODE_RESPONSE_TYPE}`);
		if (metadata.code_challenge_methods_supported && !metadata.code_challenge_methods_supported.includes(AUTHORIZATION_CODE_CHALLENGE_METHOD)) throw new Error(`Incompatible auth server: does not support code challenge method ${AUTHORIZATION_CODE_CHALLENGE_METHOD}`);
	} else authorizationUrl = new URL("/authorize", authorizationServerUrl);
	const challenge = await pkceChallenge();
	const codeVerifier = challenge.code_verifier;
	const codeChallenge = challenge.code_challenge;
	authorizationUrl.searchParams.set("response_type", AUTHORIZATION_CODE_RESPONSE_TYPE);
	authorizationUrl.searchParams.set("client_id", clientInformation.client_id);
	authorizationUrl.searchParams.set("code_challenge", codeChallenge);
	authorizationUrl.searchParams.set("code_challenge_method", AUTHORIZATION_CODE_CHALLENGE_METHOD);
	authorizationUrl.searchParams.set("redirect_uri", String(redirectUrl));
	if (state) authorizationUrl.searchParams.set("state", state);
	if (scope) authorizationUrl.searchParams.set("scope", scope);
	if (scope?.split(" ").includes("offline_access")) authorizationUrl.searchParams.append("prompt", "consent");
	if (resource) authorizationUrl.searchParams.set("resource", resource.href);
	return {
		authorizationUrl,
		codeVerifier
	};
}
/**
* Prepares token request parameters for an authorization code exchange.
*
* This is the default implementation used by {@linkcode fetchToken} when the provider
* doesn't implement {@linkcode OAuthClientProvider.prepareTokenRequest | prepareTokenRequest}.
*
* @param authorizationCode - The authorization code received from the authorization endpoint
* @param codeVerifier - The PKCE code verifier
* @param redirectUri - The redirect URI used in the authorization request
* @returns URLSearchParams for the `authorization_code` grant
*/
function prepareAuthorizationCodeRequest(authorizationCode, codeVerifier, redirectUri) {
	return new URLSearchParams({
		grant_type: "authorization_code",
		code: authorizationCode,
		code_verifier: codeVerifier,
		redirect_uri: String(redirectUri)
	});
}
/**
* Internal helper to execute a token request with the given parameters.
* Used by {@linkcode exchangeAuthorization}, {@linkcode refreshAuthorization}, and {@linkcode fetchToken}.
*/
async function executeTokenRequest(authorizationServerUrl, { metadata, tokenRequestParams, clientInformation, addClientAuthentication, resource, fetchFn }) {
	const tokenUrl = metadata?.token_endpoint ? new URL(metadata.token_endpoint) : new URL("/token", authorizationServerUrl);
	const headers = new Headers({
		"Content-Type": "application/x-www-form-urlencoded",
		Accept: "application/json"
	});
	if (resource) tokenRequestParams.set("resource", resource.href);
	if (addClientAuthentication) await addClientAuthentication(headers, tokenRequestParams, tokenUrl, metadata);
	else if (clientInformation) applyClientAuthentication(selectClientAuthMethod(clientInformation, metadata?.token_endpoint_auth_methods_supported ?? []), clientInformation, headers, tokenRequestParams);
	const response = await (fetchFn ?? fetch)(tokenUrl, {
		method: "POST",
		headers,
		body: tokenRequestParams
	});
	if (!response.ok) throw await parseErrorResponse(response);
	const json = await response.json();
	try {
		return OAuthTokensSchema.parse(json);
	} catch (parseError) {
		if (typeof json === "object" && json !== null && "error" in json) throw await parseErrorResponse(JSON.stringify(json));
		throw parseError;
	}
}
/**
* Exchanges an authorization code for an access token with the given server.
*
* Supports multiple client authentication methods as specified in OAuth 2.1:
* - Automatically selects the best authentication method based on server support
* - Falls back to appropriate defaults when server metadata is unavailable
*
* @param authorizationServerUrl - The authorization server's base URL
* @param options - Configuration object containing client info, auth code, etc.
* @returns Promise resolving to OAuth tokens
* @throws {Error} When token exchange fails or authentication is invalid
*/
async function exchangeAuthorization(authorizationServerUrl, { metadata, clientInformation, authorizationCode, codeVerifier, redirectUri, resource, addClientAuthentication, fetchFn }) {
	return executeTokenRequest(authorizationServerUrl, {
		metadata,
		tokenRequestParams: prepareAuthorizationCodeRequest(authorizationCode, codeVerifier, redirectUri),
		clientInformation,
		addClientAuthentication,
		resource,
		fetchFn
	});
}
/**
* Exchange a refresh token for an updated access token.
*
* Supports multiple client authentication methods as specified in OAuth 2.1:
* - Automatically selects the best authentication method based on server support
* - Preserves the original refresh token if a new one is not returned
*
* @param authorizationServerUrl - The authorization server's base URL
* @param options - Configuration object containing client info, refresh token, etc.
* @returns Promise resolving to OAuth tokens (preserves original `refresh_token` if not replaced)
* @throws {Error} When token refresh fails or authentication is invalid
*/
async function refreshAuthorization(authorizationServerUrl, { metadata, clientInformation, refreshToken, resource, addClientAuthentication, fetchFn }) {
	return {
		refresh_token: refreshToken,
		...await executeTokenRequest(authorizationServerUrl, {
			metadata,
			tokenRequestParams: new URLSearchParams({
				grant_type: "refresh_token",
				refresh_token: refreshToken
			}),
			clientInformation,
			addClientAuthentication,
			resource,
			fetchFn
		})
	};
}
/**
* Unified token fetching that works with any grant type via {@linkcode OAuthClientProvider.prepareTokenRequest | prepareTokenRequest()}.
*
* This function provides a single entry point for obtaining tokens regardless of the
* OAuth grant type. The provider's `prepareTokenRequest()` method determines which grant
* to use and supplies the grant-specific parameters.
*
* @param provider - OAuth client provider that implements `prepareTokenRequest()`
* @param authorizationServerUrl - The authorization server's base URL
* @param options - Configuration for the token request
* @returns Promise resolving to OAuth tokens
* @throws {Error} When provider doesn't implement `prepareTokenRequest` or token fetch fails
*
* @example
* ```ts source="./auth.examples.ts#fetchToken_clientCredentials"
* // Provider for client_credentials:
* class MyProvider extends MyProviderBase implements OAuthClientProvider {
*     prepareTokenRequest(scope?: string) {
*         const params = new URLSearchParams({ grant_type: 'client_credentials' });
*         if (scope) params.set('scope', scope);
*         return params;
*     }
* }
*
* const tokens = await fetchToken(new MyProvider(), authServerUrl, { metadata });
* ```
*/
async function fetchToken(provider, authorizationServerUrl, { metadata, resource, authorizationCode, scope, fetchFn } = {}) {
	const effectiveScope = scope ?? provider.clientMetadata.scope;
	let tokenRequestParams;
	if (provider.prepareTokenRequest) tokenRequestParams = await provider.prepareTokenRequest(effectiveScope);
	if (!tokenRequestParams) {
		if (!authorizationCode) throw new Error("Either provider.prepareTokenRequest() or authorizationCode is required");
		if (!provider.redirectUrl) throw new Error("redirectUrl is required for authorization_code flow");
		tokenRequestParams = prepareAuthorizationCodeRequest(authorizationCode, await provider.codeVerifier(), provider.redirectUrl);
	}
	const clientInformation = await provider.clientInformation();
	return executeTokenRequest(authorizationServerUrl, {
		metadata,
		tokenRequestParams,
		clientInformation: clientInformation ?? void 0,
		addClientAuthentication: provider.addClientAuthentication,
		resource,
		fetchFn
	});
}
/**
* Performs OAuth 2.0 Dynamic Client Registration according to
* {@link https://datatracker.ietf.org/doc/html/rfc7591 | RFC 7591}.
*
* If `scope` is provided, it overrides `clientMetadata.scope` in the registration
* request body. This allows callers to apply the Scope Selection Strategy (SEP-835)
* consistently across both DCR and the subsequent authorization request.
*/
async function registerClient(authorizationServerUrl, { metadata, clientMetadata, scope, fetchFn }) {
	let registrationUrl;
	if (metadata) {
		if (!metadata.registration_endpoint) throw new Error("Incompatible auth server: does not support dynamic client registration");
		registrationUrl = new URL(metadata.registration_endpoint);
	} else registrationUrl = new URL("/register", authorizationServerUrl);
	const response = await (fetchFn ?? fetch)(registrationUrl, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			...clientMetadata,
			...scope === void 0 ? {} : { scope }
		})
	});
	if (!response.ok) throw await parseErrorResponse(response);
	return OAuthClientInformationFullSchema.parse(await response.json());
}

//#endregion
//#region src/client/authExtensions.ts
/**
* Helper to produce a `private_key_jwt` client authentication function.
*
* @example
* ```ts source="./authExtensions.examples.ts#createPrivateKeyJwtAuth_basicUsage"
* const addClientAuth = createPrivateKeyJwtAuth({
*     issuer: 'my-client',
*     subject: 'my-client',
*     privateKey: pemEncodedPrivateKey,
*     alg: 'RS256'
* });
* // pass addClientAuth as provider.addClientAuthentication implementation
* ```
*/
function createPrivateKeyJwtAuth(options) {
	return async (_headers, params, url, metadata) => {
		if (globalThis.crypto === void 0) throw new TypeError("crypto is not available, please ensure you have Web Crypto API support for older Node.js versions (see https://github.com/modelcontextprotocol/typescript-sdk#nodejs-web-crypto-globalthiscrypto-compatibility)");
		const jose = await import("jose");
		const audience = String(options.audience ?? metadata?.issuer ?? url);
		const lifetimeSeconds = options.lifetimeSeconds ?? 300;
		const now = Math.floor(Date.now() / 1e3);
		const jti = `${Date.now()}-${Math.random().toString(36).slice(2)}`;
		const baseClaims = {
			iss: options.issuer,
			sub: options.subject,
			aud: audience,
			exp: now + lifetimeSeconds,
			iat: now,
			jti
		};
		const claims = options.claims ? {
			...baseClaims,
			...options.claims
		} : baseClaims;
		const alg = options.alg;
		let key;
		if (typeof options.privateKey === "string") if (alg.startsWith("RS") || alg.startsWith("ES") || alg.startsWith("PS")) key = await jose.importPKCS8(options.privateKey, alg);
		else if (alg.startsWith("HS")) key = new TextEncoder().encode(options.privateKey);
		else throw new Error(`Unsupported algorithm ${alg}`);
		else if (options.privateKey instanceof Uint8Array) key = alg.startsWith("HS") ? options.privateKey : await jose.importPKCS8(new TextDecoder().decode(options.privateKey), alg);
		else key = await jose.importJWK(options.privateKey, alg);
		const assertion = await new jose.SignJWT(claims).setProtectedHeader({
			alg,
			typ: "JWT"
		}).setIssuer(options.issuer).setSubject(options.subject).setAudience(audience).setIssuedAt(now).setExpirationTime(now + lifetimeSeconds).setJti(jti).sign(key);
		params.set("client_assertion", assertion);
		params.set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
	};
}
/**
* OAuth provider for `client_credentials` grant with `client_secret_basic` authentication.
*
* This provider is designed for machine-to-machine authentication where
* the client authenticates using a `client_id` and `client_secret`.
*
* @example
* ```ts source="./authExtensions.examples.ts#ClientCredentialsProvider_basicUsage"
* const provider = new ClientCredentialsProvider({
*     clientId: 'my-client',
*     clientSecret: 'my-secret'
* });
*
* const transport = new StreamableHTTPClientTransport(serverUrl, {
*     authProvider: provider
* });
* ```
*/
var ClientCredentialsProvider = class {
	_tokens;
	_clientInfo;
	_clientMetadata;
	constructor(options) {
		this._clientInfo = {
			client_id: options.clientId,
			client_secret: options.clientSecret
		};
		this._clientMetadata = {
			client_name: options.clientName ?? "client-credentials-client",
			redirect_uris: [],
			grant_types: ["client_credentials"],
			token_endpoint_auth_method: "client_secret_basic",
			scope: options.scope
		};
	}
	get redirectUrl() {}
	get clientMetadata() {
		return this._clientMetadata;
	}
	clientInformation() {
		return this._clientInfo;
	}
	saveClientInformation(info) {
		this._clientInfo = info;
	}
	tokens() {
		return this._tokens;
	}
	saveTokens(tokens) {
		this._tokens = tokens;
	}
	redirectToAuthorization() {
		throw new Error("redirectToAuthorization is not used for client_credentials flow");
	}
	saveCodeVerifier() {}
	codeVerifier() {
		throw new Error("codeVerifier is not used for client_credentials flow");
	}
	prepareTokenRequest(scope) {
		const params = new URLSearchParams({ grant_type: "client_credentials" });
		if (scope) params.set("scope", scope);
		return params;
	}
};
/**
* OAuth provider for `client_credentials` grant with `private_key_jwt` authentication.
*
* This provider is designed for machine-to-machine authentication where
* the client authenticates using a signed JWT assertion
* ({@link https://datatracker.ietf.org/doc/html/rfc7523#section-2.2 | RFC 7523 Section 2.2}).
*
* @example
* ```ts source="./authExtensions.examples.ts#PrivateKeyJwtProvider_basicUsage"
* const provider = new PrivateKeyJwtProvider({
*     clientId: 'my-client',
*     privateKey: pemEncodedPrivateKey,
*     algorithm: 'RS256'
* });
*
* const transport = new StreamableHTTPClientTransport(serverUrl, {
*     authProvider: provider
* });
* ```
*/
var PrivateKeyJwtProvider = class {
	_tokens;
	_clientInfo;
	_clientMetadata;
	addClientAuthentication;
	constructor(options) {
		this._clientInfo = { client_id: options.clientId };
		this._clientMetadata = {
			client_name: options.clientName ?? "private-key-jwt-client",
			redirect_uris: [],
			grant_types: ["client_credentials"],
			token_endpoint_auth_method: "private_key_jwt",
			scope: options.scope
		};
		this.addClientAuthentication = createPrivateKeyJwtAuth({
			issuer: options.clientId,
			subject: options.clientId,
			privateKey: options.privateKey,
			alg: options.algorithm,
			lifetimeSeconds: options.jwtLifetimeSeconds
		});
	}
	get redirectUrl() {}
	get clientMetadata() {
		return this._clientMetadata;
	}
	clientInformation() {
		return this._clientInfo;
	}
	saveClientInformation(info) {
		this._clientInfo = info;
	}
	tokens() {
		return this._tokens;
	}
	saveTokens(tokens) {
		this._tokens = tokens;
	}
	redirectToAuthorization() {
		throw new Error("redirectToAuthorization is not used for client_credentials flow");
	}
	saveCodeVerifier() {}
	codeVerifier() {
		throw new Error("codeVerifier is not used for client_credentials flow");
	}
	prepareTokenRequest(scope) {
		const params = new URLSearchParams({ grant_type: "client_credentials" });
		if (scope) params.set("scope", scope);
		return params;
	}
};
/**
* OAuth provider for `client_credentials` grant with a static `private_key_jwt` assertion.
*
* This provider mirrors {@linkcode PrivateKeyJwtProvider} but instead of constructing and
* signing a JWT on each request, it accepts a pre-built JWT assertion string and
* uses it directly for authentication.
*/
var StaticPrivateKeyJwtProvider = class {
	_tokens;
	_clientInfo;
	_clientMetadata;
	addClientAuthentication;
	constructor(options) {
		this._clientInfo = { client_id: options.clientId };
		this._clientMetadata = {
			client_name: options.clientName ?? "static-private-key-jwt-client",
			redirect_uris: [],
			grant_types: ["client_credentials"],
			token_endpoint_auth_method: "private_key_jwt",
			scope: options.scope
		};
		const assertion = options.jwtBearerAssertion;
		this.addClientAuthentication = async (_headers, params) => {
			params.set("client_assertion", assertion);
			params.set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
		};
	}
	get redirectUrl() {}
	get clientMetadata() {
		return this._clientMetadata;
	}
	clientInformation() {
		return this._clientInfo;
	}
	saveClientInformation(info) {
		this._clientInfo = info;
	}
	tokens() {
		return this._tokens;
	}
	saveTokens(tokens) {
		this._tokens = tokens;
	}
	redirectToAuthorization() {
		throw new Error("redirectToAuthorization is not used for client_credentials flow");
	}
	saveCodeVerifier() {}
	codeVerifier() {
		throw new Error("codeVerifier is not used for client_credentials flow");
	}
	prepareTokenRequest(scope) {
		const params = new URLSearchParams({ grant_type: "client_credentials" });
		if (scope) params.set("scope", scope);
		return params;
	}
};
/**
* OAuth provider for Cross-App Access (Enterprise Managed Authorization) using JWT Authorization Grant.
*
* This provider implements the Enterprise Managed Authorization flow (SEP-990) where:
* 1. User authenticates with an enterprise IdP and the client obtains an ID Token
* 2. Client exchanges the ID Token for a JWT Authorization Grant (ID-JAG) via RFC 8693 token exchange
* 3. Client uses the JAG to obtain an access token from the MCP server via RFC 7523 JWT bearer grant
*
* The provider handles steps 2-3 automatically, with the JAG acquisition delegated to
* a callback function that you provide. This allows flexibility in how you obtain and
* cache ID Tokens from the IdP.
*
* @see https://github.com/modelcontextprotocol/ext-auth/blob/main/specification/draft/enterprise-managed-authorization.mdx
*
* @example
* ```ts
* const provider = new CrossAppAccessProvider({
*     assertion: async (ctx) => {
*         const result = await discoverAndRequestJwtAuthGrant({
*             idpUrl: 'https://idp.example.com',
*             audience: ctx.authorizationServerUrl,
*             resource: ctx.resourceUrl,
*             idToken: await getIdToken(), // Your function to get ID token
*             clientId: 'my-idp-client',
*             clientSecret: 'my-idp-secret',
*             scope: ctx.scope,
*             fetchFn: ctx.fetchFn
*         });
*         return result.jwtAuthGrant;
*     },
*     clientId: 'my-mcp-client',
*     clientSecret: 'my-mcp-secret'
* });
*
* const transport = new StreamableHTTPClientTransport(serverUrl, {
*     authProvider: provider
* });
* ```
*/
var CrossAppAccessProvider = class {
	_tokens;
	_clientInfo;
	_clientMetadata;
	_assertionCallback;
	_fetchFn;
	_authorizationServerUrl;
	_resourceUrl;
	_scope;
	constructor(options) {
		this._clientInfo = {
			client_id: options.clientId,
			client_secret: options.clientSecret
		};
		this._clientMetadata = {
			client_name: options.clientName ?? "cross-app-access-client",
			redirect_uris: [],
			grant_types: ["urn:ietf:params:oauth:grant-type:jwt-bearer"],
			token_endpoint_auth_method: "client_secret_basic"
		};
		this._assertionCallback = options.assertion;
		this._fetchFn = options.fetchFn ?? fetch;
	}
	get redirectUrl() {}
	get clientMetadata() {
		return this._clientMetadata;
	}
	clientInformation() {
		return this._clientInfo;
	}
	saveClientInformation(info) {
		this._clientInfo = info;
	}
	tokens() {
		return this._tokens;
	}
	saveTokens(tokens) {
		this._tokens = tokens;
	}
	redirectToAuthorization() {
		throw new Error("redirectToAuthorization is not used for jwt-bearer flow");
	}
	saveCodeVerifier() {}
	codeVerifier() {
		throw new Error("codeVerifier is not used for jwt-bearer flow");
	}
	/**
	* Saves the authorization server URL discovered during OAuth flow.
	* This is called by the auth() function after RFC 9728 discovery.
	*/
	saveAuthorizationServerUrl(authorizationServerUrl) {
		this._authorizationServerUrl = authorizationServerUrl;
	}
	/**
	* Returns the cached authorization server URL if available.
	*/
	authorizationServerUrl() {
		return this._authorizationServerUrl;
	}
	/**
	* Saves the resource URL discovered during OAuth flow.
	* This is called by the auth() function after RFC 9728 discovery.
	*/
	saveResourceUrl(resourceUrl) {
		this._resourceUrl = resourceUrl;
	}
	/**
	* Returns the cached resource URL if available.
	*/
	resourceUrl() {
		return this._resourceUrl;
	}
	async prepareTokenRequest(scope) {
		const authServerUrl = this._authorizationServerUrl;
		const resourceUrl = this._resourceUrl;
		if (!authServerUrl) throw new Error("Authorization server URL not available. Ensure auth() has been called first.");
		if (!resourceUrl) throw new Error("Resource URL not available — server may not implement RFC 9728 Protected Resource Metadata (required for Cross-App Access), or auth() has not been called");
		this._scope = scope;
		const jwtAuthGrant = await this._assertionCallback({
			authorizationServerUrl: authServerUrl,
			resourceUrl,
			scope: this._scope,
			fetchFn: this._fetchFn
		});
		const params = new URLSearchParams({
			grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
			assertion: jwtAuthGrant
		});
		if (scope) params.set("scope", scope);
		return params;
	}
};

//#endregion
//#region src/experimental/tasks/client.ts
/**
* Experimental task features for MCP clients.
*
* Access via `client.experimental.tasks`:
* ```typescript
* const stream = client.experimental.tasks.callToolStream({ name: 'tool', arguments: {} });
* const task = await client.experimental.tasks.getTask(taskId);
* ```
*
* @experimental
*/
var ExperimentalClientTasks = class {
	constructor(_client) {
		this._client = _client;
	}
	get _module() {
		return this._client.taskManager;
	}
	/**
	* Calls a tool and returns an AsyncGenerator that yields response messages.
	* The generator is guaranteed to end with either a `'result'` or `'error'` message.
	*
	* This method provides streaming access to tool execution, allowing you to
	* observe intermediate task status updates for long-running tool calls.
	* Automatically validates structured output if the tool has an `outputSchema`.
	*
	* @example
	* ```ts source="./client.examples.ts#ExperimentalClientTasks_callToolStream"
	* const stream = client.experimental.tasks.callToolStream({ name: 'myTool', arguments: {} });
	* for await (const message of stream) {
	*     switch (message.type) {
	*         case 'taskCreated': {
	*             console.log('Tool execution started:', message.task.taskId);
	*             break;
	*         }
	*         case 'taskStatus': {
	*             console.log('Tool status:', message.task.status);
	*             break;
	*         }
	*         case 'result': {
	*             console.log('Tool result:', message.result);
	*             break;
	*         }
	*         case 'error': {
	*             console.error('Tool error:', message.error);
	*             break;
	*         }
	*     }
	* }
	* ```
	*
	* @param params - Tool call parameters (name and arguments)
	* @param options - Optional request options (timeout, signal, task creation params, etc.)
	* @returns AsyncGenerator that yields {@linkcode ResponseMessage} objects
	*
	* @experimental
	*/
	async *callToolStream(params, options) {
		const clientInternal = this._client;
		const optionsWithTask = {
			...options,
			task: options?.task ?? (clientInternal.isToolTask(params.name) ? {} : void 0)
		};
		const stream = this._module.requestStream({
			method: "tools/call",
			params
		}, CallToolResultSchema, optionsWithTask);
		const validator = clientInternal.getToolOutputValidator(params.name);
		for await (const message of stream) {
			if (message.type === "result" && validator && "content" in message.result) {
				const result = message.result;
				if (!result.structuredContent && !result.isError) {
					yield {
						type: "error",
						error: new ProtocolError(ProtocolErrorCode.InvalidRequest, `Tool ${params.name} has an output schema but did not return structured content`)
					};
					return;
				}
				if (result.structuredContent) try {
					const validationResult = validator(result.structuredContent);
					if (!validationResult.valid) {
						yield {
							type: "error",
							error: new ProtocolError(ProtocolErrorCode.InvalidParams, `Structured content does not match the tool's output schema: ${validationResult.errorMessage}`)
						};
						return;
					}
				} catch (error) {
					if (error instanceof ProtocolError) {
						yield {
							type: "error",
							error
						};
						return;
					}
					yield {
						type: "error",
						error: new ProtocolError(ProtocolErrorCode.InvalidParams, `Failed to validate structured content: ${error instanceof Error ? error.message : String(error)}`)
					};
					return;
				}
			}
			yield message;
		}
	}
	/**
	* Gets the current status of a task.
	*
	* @param taskId - The task identifier
	* @param options - Optional request options
	* @returns The task status
	*
	* @experimental
	*/
	async getTask(taskId, options) {
		return this._module.getTask({ taskId }, options);
	}
	/**
	* Retrieves the result of a completed task.
	*
	* @param taskId - The task identifier
	* @param options - Optional request options
	* @returns The task result. The payload structure matches the result type of the
	*   original request (e.g., a `tools/call` task returns a `CallToolResult`).
	*
	* @experimental
	*/
	async getTaskResult(taskId, options) {
		return this._module.getTaskResult({ taskId }, GetTaskPayloadResultSchema, options);
	}
	/**
	* Lists tasks with optional pagination.
	*
	* @param cursor - Optional pagination cursor
	* @param options - Optional request options
	* @returns List of tasks with optional next cursor
	*
	* @experimental
	*/
	async listTasks(cursor, options) {
		return this._module.listTasks(cursor ? { cursor } : void 0, options);
	}
	/**
	* Cancels a running task.
	*
	* @param taskId - The task identifier
	* @param options - Optional request options
	*
	* @experimental
	*/
	async cancelTask(taskId, options) {
		return this._module.cancelTask({ taskId }, options);
	}
	/**
	* Sends a request and returns an AsyncGenerator that yields response messages.
	* The generator is guaranteed to end with either a `'result'` or `'error'` message.
	*
	* This method provides streaming access to request processing, allowing you to
	* observe intermediate task status updates for task-augmented requests.
	*
	* @example
	* ```ts source="./client.examples.ts#ExperimentalClientTasks_requestStream"
	* const stream = client.experimental.tasks.requestStream({ method: 'tools/call', params: { name: 'my-tool', arguments: {} } }, options);
	* for await (const message of stream) {
	*     switch (message.type) {
	*         case 'taskCreated': {
	*             console.log('Task created:', message.task.taskId);
	*             break;
	*         }
	*         case 'taskStatus': {
	*             console.log('Task status:', message.task.status);
	*             break;
	*         }
	*         case 'result': {
	*             console.log('Final result:', message.result);
	*             break;
	*         }
	*         case 'error': {
	*             console.error('Error:', message.error);
	*             break;
	*         }
	*     }
	* }
	* ```
	*
	* @param request - The request to send
	* @param options - Optional request options (timeout, signal, task creation params, etc.)
	* @returns AsyncGenerator that yields {@linkcode ResponseMessage} objects
	*
	* @experimental
	*/
	requestStream(request, options) {
		const resultSchema = getResultSchema(request.method);
		return this._module.requestStream(request, resultSchema, options);
	}
};

//#endregion
//#region src/client/client.ts
/**
* Elicitation default application helper. Applies defaults to the `data` based on the `schema`.
*
* @param schema - The schema to apply defaults to.
* @param data - The data to apply defaults to.
*/
function applyElicitationDefaults(schema, data) {
	if (!schema || data === null || typeof data !== "object") return;
	if (schema.type === "object" && schema.properties && typeof schema.properties === "object") {
		const obj = data;
		const props = schema.properties;
		for (const key of Object.keys(props)) {
			const propSchema = props[key];
			if (obj[key] === void 0 && Object.prototype.hasOwnProperty.call(propSchema, "default")) obj[key] = propSchema.default;
			if (obj[key] !== void 0) applyElicitationDefaults(propSchema, obj[key]);
		}
	}
	if (Array.isArray(schema.anyOf)) {
		for (const sub of schema.anyOf) if (typeof sub !== "boolean") applyElicitationDefaults(sub, data);
	}
	if (Array.isArray(schema.oneOf)) {
		for (const sub of schema.oneOf) if (typeof sub !== "boolean") applyElicitationDefaults(sub, data);
	}
}
/**
* Determines which elicitation modes are supported based on declared client capabilities.
*
* According to the spec:
* - An empty elicitation capability object defaults to form mode support (backwards compatibility)
* - URL mode is only supported if explicitly declared
*
* @param capabilities - The client's elicitation capabilities
* @returns An object indicating which modes are supported
*/
function getSupportedElicitationModes(capabilities) {
	if (!capabilities) return {
		supportsFormMode: false,
		supportsUrlMode: false
	};
	const hasFormCapability = capabilities.form !== void 0;
	const hasUrlCapability = capabilities.url !== void 0;
	return {
		supportsFormMode: hasFormCapability || !hasFormCapability && !hasUrlCapability,
		supportsUrlMode: hasUrlCapability
	};
}
/**
* An MCP client on top of a pluggable transport.
*
* The client will automatically begin the initialization flow with the server when {@linkcode connect} is called.
*
*/
var Client = class extends Protocol {
	_serverCapabilities;
	_serverVersion;
	_negotiatedProtocolVersion;
	_capabilities;
	_instructions;
	_jsonSchemaValidator;
	_cachedToolOutputValidators = /* @__PURE__ */ new Map();
	_cachedKnownTaskTools = /* @__PURE__ */ new Set();
	_cachedRequiredTaskTools = /* @__PURE__ */ new Set();
	_experimental;
	_listChangedDebounceTimers = /* @__PURE__ */ new Map();
	_pendingListChangedConfig;
	_enforceStrictCapabilities;
	/**
	* Initializes this client with the given name and version information.
	*/
	constructor(_clientInfo, options) {
		super({
			...options,
			tasks: extractTaskManagerOptions(options?.capabilities?.tasks)
		});
		this._clientInfo = _clientInfo;
		this._capabilities = options?.capabilities ? { ...options.capabilities } : {};
		this._jsonSchemaValidator = options?.jsonSchemaValidator ?? new DefaultJsonSchemaValidator();
		this._enforceStrictCapabilities = options?.enforceStrictCapabilities ?? false;
		if (options?.capabilities?.tasks) {
			const { taskStore, taskMessageQueue, defaultTaskPollInterval, maxTaskQueueSize, ...wireCapabilities } = options.capabilities.tasks;
			this._capabilities.tasks = wireCapabilities;
		}
		if (options?.listChanged) this._pendingListChangedConfig = options.listChanged;
	}
	buildContext(ctx, _transportInfo) {
		return ctx;
	}
	/**
	* Set up handlers for list changed notifications based on config and server capabilities.
	* This should only be called after initialization when server capabilities are known.
	* Handlers are silently skipped if the server doesn't advertise the corresponding listChanged capability.
	* @internal
	*/
	_setupListChangedHandlers(config) {
		if (config.tools && this._serverCapabilities?.tools?.listChanged) this._setupListChangedHandler("tools", "notifications/tools/list_changed", config.tools, async () => {
			return (await this.listTools()).tools;
		});
		if (config.prompts && this._serverCapabilities?.prompts?.listChanged) this._setupListChangedHandler("prompts", "notifications/prompts/list_changed", config.prompts, async () => {
			return (await this.listPrompts()).prompts;
		});
		if (config.resources && this._serverCapabilities?.resources?.listChanged) this._setupListChangedHandler("resources", "notifications/resources/list_changed", config.resources, async () => {
			return (await this.listResources()).resources;
		});
	}
	/**
	* Access experimental features.
	*
	* WARNING: These APIs are experimental and may change without notice.
	*
	* @experimental
	*/
	get experimental() {
		if (!this._experimental) this._experimental = { tasks: new ExperimentalClientTasks(this) };
		return this._experimental;
	}
	/**
	* Registers new capabilities. This can only be called before connecting to a transport.
	*
	* The new capabilities will be merged with any existing capabilities previously given (e.g., at initialization).
	*/
	registerCapabilities(capabilities) {
		if (this.transport) throw new Error("Cannot register capabilities after connecting to transport");
		this._capabilities = mergeCapabilities(this._capabilities, capabilities);
	}
	/**
	* Registers a handler for server-initiated requests (sampling, elicitation, roots).
	* The client must declare the corresponding capability for the handler to be accepted.
	* Replaces any previously registered handler for the same method.
	*
	* For `sampling/createMessage` and `elicitation/create`, the handler is automatically
	* wrapped with schema validation for both the incoming request and the returned result.
	*
	* @example Handling a sampling request
	* ```ts source="./client.examples.ts#Client_setRequestHandler_sampling"
	* client.setRequestHandler('sampling/createMessage', async request => {
	*     const lastMessage = request.params.messages.at(-1);
	*     console.log('Sampling request:', lastMessage);
	*
	*     // In production, send messages to your LLM here
	*     return {
	*         model: 'my-model',
	*         role: 'assistant' as const,
	*         content: {
	*             type: 'text' as const,
	*             text: 'Response from the model'
	*         }
	*     };
	* });
	* ```
	*/
	setRequestHandler(method, handler) {
		if (method === "elicitation/create") {
			const wrappedHandler = async (request, ctx) => {
				const validatedRequest = parseSchema(ElicitRequestSchema, request);
				if (!validatedRequest.success) {
					const errorMessage = validatedRequest.error instanceof Error ? validatedRequest.error.message : String(validatedRequest.error);
					throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Invalid elicitation request: ${errorMessage}`);
				}
				const { params } = validatedRequest.data;
				params.mode = params.mode ?? "form";
				const { supportsFormMode, supportsUrlMode } = getSupportedElicitationModes(this._capabilities.elicitation);
				if (params.mode === "form" && !supportsFormMode) throw new ProtocolError(ProtocolErrorCode.InvalidParams, "Client does not support form-mode elicitation requests");
				if (params.mode === "url" && !supportsUrlMode) throw new ProtocolError(ProtocolErrorCode.InvalidParams, "Client does not support URL-mode elicitation requests");
				const result = await Promise.resolve(handler(request, ctx));
				if (params.task) {
					const taskValidationResult = parseSchema(CreateTaskResultSchema, result);
					if (!taskValidationResult.success) {
						const errorMessage = taskValidationResult.error instanceof Error ? taskValidationResult.error.message : String(taskValidationResult.error);
						throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Invalid task creation result: ${errorMessage}`);
					}
					return taskValidationResult.data;
				}
				const validationResult = parseSchema(ElicitResultSchema, result);
				if (!validationResult.success) {
					const errorMessage = validationResult.error instanceof Error ? validationResult.error.message : String(validationResult.error);
					throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Invalid elicitation result: ${errorMessage}`);
				}
				const validatedResult = validationResult.data;
				const requestedSchema = params.mode === "form" ? params.requestedSchema : void 0;
				if (params.mode === "form" && validatedResult.action === "accept" && validatedResult.content && requestedSchema && this._capabilities.elicitation?.form?.applyDefaults) try {
					applyElicitationDefaults(requestedSchema, validatedResult.content);
				} catch {}
				return validatedResult;
			};
			return super.setRequestHandler(method, wrappedHandler);
		}
		if (method === "sampling/createMessage") {
			const wrappedHandler = async (request, ctx) => {
				const validatedRequest = parseSchema(CreateMessageRequestSchema, request);
				if (!validatedRequest.success) {
					const errorMessage = validatedRequest.error instanceof Error ? validatedRequest.error.message : String(validatedRequest.error);
					throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Invalid sampling request: ${errorMessage}`);
				}
				const { params } = validatedRequest.data;
				const result = await Promise.resolve(handler(request, ctx));
				if (params.task) {
					const taskValidationResult = parseSchema(CreateTaskResultSchema, result);
					if (!taskValidationResult.success) {
						const errorMessage = taskValidationResult.error instanceof Error ? taskValidationResult.error.message : String(taskValidationResult.error);
						throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Invalid task creation result: ${errorMessage}`);
					}
					return taskValidationResult.data;
				}
				const validationResult = parseSchema(params.tools || params.toolChoice ? CreateMessageResultWithToolsSchema : CreateMessageResultSchema, result);
				if (!validationResult.success) {
					const errorMessage = validationResult.error instanceof Error ? validationResult.error.message : String(validationResult.error);
					throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Invalid sampling result: ${errorMessage}`);
				}
				return validationResult.data;
			};
			return super.setRequestHandler(method, wrappedHandler);
		}
		return super.setRequestHandler(method, handler);
	}
	assertCapability(capability, method) {
		if (!this._serverCapabilities?.[capability]) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support ${capability} (required for ${method})`);
	}
	/**
	* Connects to a server via the given transport and performs the MCP initialization handshake.
	*
	* @example Basic usage (stdio)
	* ```ts source="./client.examples.ts#Client_connect_stdio"
	* const client = new Client({ name: 'my-client', version: '1.0.0' });
	* const transport = new StdioClientTransport({ command: 'my-mcp-server' });
	* await client.connect(transport);
	* ```
	*
	* @example Streamable HTTP with SSE fallback
	* ```ts source="./client.examples.ts#Client_connect_sseFallback"
	* const baseUrl = new URL(url);
	*
	* try {
	*     // Try modern Streamable HTTP transport first
	*     const client = new Client({ name: 'my-client', version: '1.0.0' });
	*     const transport = new StreamableHTTPClientTransport(baseUrl);
	*     await client.connect(transport);
	*     return { client, transport };
	* } catch {
	*     // Fall back to legacy SSE transport
	*     const client = new Client({ name: 'my-client', version: '1.0.0' });
	*     const transport = new SSEClientTransport(baseUrl);
	*     await client.connect(transport);
	*     return { client, transport };
	* }
	* ```
	*/
	async connect(transport, options) {
		await super.connect(transport);
		if (transport.sessionId !== void 0) {
			if (this._negotiatedProtocolVersion !== void 0 && transport.setProtocolVersion) transport.setProtocolVersion(this._negotiatedProtocolVersion);
			return;
		}
		try {
			const result = await this._requestWithSchema({
				method: "initialize",
				params: {
					protocolVersion: this._supportedProtocolVersions[0] ?? LATEST_PROTOCOL_VERSION,
					capabilities: this._capabilities,
					clientInfo: this._clientInfo
				}
			}, InitializeResultSchema, options);
			if (result === void 0) throw new Error(`Server sent invalid initialize result: ${result}`);
			if (!this._supportedProtocolVersions.includes(result.protocolVersion)) throw new Error(`Server's protocol version is not supported: ${result.protocolVersion}`);
			this._serverCapabilities = result.capabilities;
			this._serverVersion = result.serverInfo;
			this._negotiatedProtocolVersion = result.protocolVersion;
			if (transport.setProtocolVersion) transport.setProtocolVersion(result.protocolVersion);
			this._instructions = result.instructions;
			await this.notification({ method: "notifications/initialized" });
			if (this._pendingListChangedConfig) {
				this._setupListChangedHandlers(this._pendingListChangedConfig);
				this._pendingListChangedConfig = void 0;
			}
		} catch (error) {
			this.close();
			throw error;
		}
	}
	/**
	* After initialization has completed, this will be populated with the server's reported capabilities.
	*/
	getServerCapabilities() {
		return this._serverCapabilities;
	}
	/**
	* After initialization has completed, this will be populated with information about the server's name and version.
	*/
	getServerVersion() {
		return this._serverVersion;
	}
	/**
	* After initialization has completed, this will be populated with the protocol version negotiated
	* during the initialize handshake. When manually reconstructing a transport for reconnection, pass this
	* value to the new transport so it continues sending the required `mcp-protocol-version` header.
	*/
	getNegotiatedProtocolVersion() {
		return this._negotiatedProtocolVersion;
	}
	/**
	* After initialization has completed, this may be populated with information about the server's instructions.
	*/
	getInstructions() {
		return this._instructions;
	}
	assertCapabilityForMethod(method) {
		switch (method) {
			case "logging/setLevel":
				if (!this._serverCapabilities?.logging) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support logging (required for ${method})`);
				break;
			case "prompts/get":
			case "prompts/list":
				if (!this._serverCapabilities?.prompts) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support prompts (required for ${method})`);
				break;
			case "resources/list":
			case "resources/templates/list":
			case "resources/read":
			case "resources/subscribe":
			case "resources/unsubscribe":
				if (!this._serverCapabilities?.resources) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support resources (required for ${method})`);
				if (method === "resources/subscribe" && !this._serverCapabilities.resources.subscribe) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support resource subscriptions (required for ${method})`);
				break;
			case "tools/call":
			case "tools/list":
				if (!this._serverCapabilities?.tools) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support tools (required for ${method})`);
				break;
			case "completion/complete":
				if (!this._serverCapabilities?.completions) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Server does not support completions (required for ${method})`);
				break;
			case "initialize": break;
			case "ping": break;
		}
	}
	assertNotificationCapability(method) {
		switch (method) {
			case "notifications/roots/list_changed":
				if (!this._capabilities.roots?.listChanged) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Client does not support roots list changed notifications (required for ${method})`);
				break;
			case "notifications/initialized": break;
			case "notifications/cancelled": break;
			case "notifications/progress": break;
		}
	}
	assertRequestHandlerCapability(method) {
		switch (method) {
			case "sampling/createMessage":
				if (!this._capabilities.sampling) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Client does not support sampling capability (required for ${method})`);
				break;
			case "elicitation/create":
				if (!this._capabilities.elicitation) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Client does not support elicitation capability (required for ${method})`);
				break;
			case "roots/list":
				if (!this._capabilities.roots) throw new SdkError(SdkErrorCode.CapabilityNotSupported, `Client does not support roots capability (required for ${method})`);
				break;
			case "ping": break;
		}
	}
	assertTaskCapability(method) {
		assertToolsCallTaskCapability(this._serverCapabilities?.tasks?.requests, method, "Server");
	}
	assertTaskHandlerCapability(method) {
		assertClientRequestTaskCapability(this._capabilities?.tasks?.requests, method, "Client");
	}
	async ping(options) {
		return this._requestWithSchema({ method: "ping" }, EmptyResultSchema, options);
	}
	/** Requests argument autocompletion suggestions from the server for a prompt or resource. */
	async complete(params, options) {
		return this._requestWithSchema({
			method: "completion/complete",
			params
		}, CompleteResultSchema, options);
	}
	/** Sets the minimum severity level for log messages sent by the server. */
	async setLoggingLevel(level, options) {
		return this._requestWithSchema({
			method: "logging/setLevel",
			params: { level }
		}, EmptyResultSchema, options);
	}
	/** Retrieves a prompt by name from the server, passing the given arguments for template substitution. */
	async getPrompt(params, options) {
		return this._requestWithSchema({
			method: "prompts/get",
			params
		}, GetPromptResultSchema, options);
	}
	/**
	* Lists available prompts. Results may be paginated — loop on `nextCursor` to collect all pages.
	*
	* Returns an empty list if the server does not advertise prompts capability
	* (or throws if {@linkcode ClientOptions.enforceStrictCapabilities} is enabled).
	*
	* @example
	* ```ts source="./client.examples.ts#Client_listPrompts_pagination"
	* const allPrompts: Prompt[] = [];
	* let cursor: string | undefined;
	* do {
	*     const { prompts, nextCursor } = await client.listPrompts({ cursor });
	*     allPrompts.push(...prompts);
	*     cursor = nextCursor;
	* } while (cursor);
	* console.log(
	*     'Available prompts:',
	*     allPrompts.map(p => p.name)
	* );
	* ```
	*/
	async listPrompts(params, options) {
		if (!this._serverCapabilities?.prompts && !this._enforceStrictCapabilities) {
			console.debug("Client.listPrompts() called but server does not advertise prompts capability - returning empty list");
			return { prompts: [] };
		}
		return this._requestWithSchema({
			method: "prompts/list",
			params
		}, ListPromptsResultSchema, options);
	}
	/**
	* Lists available resources. Results may be paginated — loop on `nextCursor` to collect all pages.
	*
	* Returns an empty list if the server does not advertise resources capability
	* (or throws if {@linkcode ClientOptions.enforceStrictCapabilities} is enabled).
	*
	* @example
	* ```ts source="./client.examples.ts#Client_listResources_pagination"
	* const allResources: Resource[] = [];
	* let cursor: string | undefined;
	* do {
	*     const { resources, nextCursor } = await client.listResources({ cursor });
	*     allResources.push(...resources);
	*     cursor = nextCursor;
	* } while (cursor);
	* console.log(
	*     'Available resources:',
	*     allResources.map(r => r.name)
	* );
	* ```
	*/
	async listResources(params, options) {
		if (!this._serverCapabilities?.resources && !this._enforceStrictCapabilities) {
			console.debug("Client.listResources() called but server does not advertise resources capability - returning empty list");
			return { resources: [] };
		}
		return this._requestWithSchema({
			method: "resources/list",
			params
		}, ListResourcesResultSchema, options);
	}
	/**
	* Lists available resource URI templates for dynamic resources. Results may be paginated — see {@linkcode listResources | listResources()} for the cursor pattern.
	*
	* Returns an empty list if the server does not advertise resources capability
	* (or throws if {@linkcode ClientOptions.enforceStrictCapabilities} is enabled).
	*/
	async listResourceTemplates(params, options) {
		if (!this._serverCapabilities?.resources && !this._enforceStrictCapabilities) {
			console.debug("Client.listResourceTemplates() called but server does not advertise resources capability - returning empty list");
			return { resourceTemplates: [] };
		}
		return this._requestWithSchema({
			method: "resources/templates/list",
			params
		}, ListResourceTemplatesResultSchema, options);
	}
	/** Reads the contents of a resource by URI. */
	async readResource(params, options) {
		return this._requestWithSchema({
			method: "resources/read",
			params
		}, ReadResourceResultSchema, options);
	}
	/** Subscribes to change notifications for a resource. The server must support resource subscriptions. */
	async subscribeResource(params, options) {
		return this._requestWithSchema({
			method: "resources/subscribe",
			params
		}, EmptyResultSchema, options);
	}
	/** Unsubscribes from change notifications for a resource. */
	async unsubscribeResource(params, options) {
		return this._requestWithSchema({
			method: "resources/unsubscribe",
			params
		}, EmptyResultSchema, options);
	}
	/**
	* Calls a tool on the connected server and returns the result. Automatically validates structured output
	* if the tool has an `outputSchema`.
	*
	* Tool results have two error surfaces: `result.isError` for tool-level failures (the tool ran but reported
	* a problem), and thrown {@linkcode ProtocolError} for protocol-level failures or {@linkcode SdkError} for
	* SDK-level issues (timeouts, missing capabilities).
	*
	* For task-based execution with streaming behavior, use {@linkcode ExperimentalClientTasks.callToolStream | client.experimental.tasks.callToolStream()} instead.
	*
	* @example Basic usage
	* ```ts source="./client.examples.ts#Client_callTool_basic"
	* const result = await client.callTool({
	*     name: 'calculate-bmi',
	*     arguments: { weightKg: 70, heightM: 1.75 }
	* });
	*
	* // Tool-level errors are returned in the result, not thrown
	* if (result.isError) {
	*     console.error('Tool error:', result.content);
	*     return;
	* }
	*
	* console.log(result.content);
	* ```
	*
	* @example Structured output
	* ```ts source="./client.examples.ts#Client_callTool_structuredOutput"
	* const result = await client.callTool({
	*     name: 'calculate-bmi',
	*     arguments: { weightKg: 70, heightM: 1.75 }
	* });
	*
	* // Machine-readable output for the client application
	* if (result.structuredContent) {
	*     console.log(result.structuredContent); // e.g. { bmi: 22.86 }
	* }
	* ```
	*/
	async callTool(params, options) {
		if (this.isToolTaskRequired(params.name)) throw new ProtocolError(ProtocolErrorCode.InvalidRequest, `Tool "${params.name}" requires task-based execution. Use client.experimental.tasks.callToolStream() instead.`);
		const result = await this._requestWithSchema({
			method: "tools/call",
			params
		}, CallToolResultSchema, options);
		const validator = this.getToolOutputValidator(params.name);
		if (validator) {
			if (!result.structuredContent && !result.isError) throw new ProtocolError(ProtocolErrorCode.InvalidRequest, `Tool ${params.name} has an output schema but did not return structured content`);
			if (result.structuredContent) try {
				const validationResult = validator(result.structuredContent);
				if (!validationResult.valid) throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Structured content does not match the tool's output schema: ${validationResult.errorMessage}`);
			} catch (error) {
				if (error instanceof ProtocolError) throw error;
				throw new ProtocolError(ProtocolErrorCode.InvalidParams, `Failed to validate structured content: ${error instanceof Error ? error.message : String(error)}`);
			}
		}
		return result;
	}
	isToolTask(toolName) {
		if (!this._serverCapabilities?.tasks?.requests?.tools?.call) return false;
		return this._cachedKnownTaskTools.has(toolName);
	}
	/**
	* Check if a tool requires task-based execution.
	* Unlike {@linkcode isToolTask} which includes `'optional'` tools, this only checks for `'required'`.
	*/
	isToolTaskRequired(toolName) {
		return this._cachedRequiredTaskTools.has(toolName);
	}
	/**
	* Cache validators for tool output schemas.
	* Called after {@linkcode listTools | listTools()} to pre-compile validators for better performance.
	*/
	cacheToolMetadata(tools) {
		this._cachedToolOutputValidators.clear();
		this._cachedKnownTaskTools.clear();
		this._cachedRequiredTaskTools.clear();
		for (const tool of tools) {
			if (tool.outputSchema) {
				const toolValidator = this._jsonSchemaValidator.getValidator(tool.outputSchema);
				this._cachedToolOutputValidators.set(tool.name, toolValidator);
			}
			const taskSupport = tool.execution?.taskSupport;
			if (taskSupport === "required" || taskSupport === "optional") this._cachedKnownTaskTools.add(tool.name);
			if (taskSupport === "required") this._cachedRequiredTaskTools.add(tool.name);
		}
	}
	/**
	* Get cached validator for a tool
	*/
	getToolOutputValidator(toolName) {
		return this._cachedToolOutputValidators.get(toolName);
	}
	/**
	* Lists available tools. Results may be paginated — loop on `nextCursor` to collect all pages.
	*
	* Returns an empty list if the server does not advertise tools capability
	* (or throws if {@linkcode ClientOptions.enforceStrictCapabilities} is enabled).
	*
	* @example
	* ```ts source="./client.examples.ts#Client_listTools_pagination"
	* const allTools: Tool[] = [];
	* let cursor: string | undefined;
	* do {
	*     const { tools, nextCursor } = await client.listTools({ cursor });
	*     allTools.push(...tools);
	*     cursor = nextCursor;
	* } while (cursor);
	* console.log(
	*     'Available tools:',
	*     allTools.map(t => t.name)
	* );
	* ```
	*/
	async listTools(params, options) {
		if (!this._serverCapabilities?.tools && !this._enforceStrictCapabilities) {
			console.debug("Client.listTools() called but server does not advertise tools capability - returning empty list");
			return { tools: [] };
		}
		const result = await this._requestWithSchema({
			method: "tools/list",
			params
		}, ListToolsResultSchema, options);
		this.cacheToolMetadata(result.tools);
		return result;
	}
	/**
	* Set up a single list changed handler.
	* @internal
	*/
	_setupListChangedHandler(listType, notificationMethod, options, fetcher) {
		const parseResult = parseSchema(ListChangedOptionsBaseSchema, options);
		if (!parseResult.success) throw new Error(`Invalid ${listType} listChanged options: ${parseResult.error.message}`);
		if (typeof options.onChanged !== "function") throw new TypeError(`Invalid ${listType} listChanged options: onChanged must be a function`);
		const { autoRefresh, debounceMs } = parseResult.data;
		const { onChanged } = options;
		const refresh = async () => {
			if (!autoRefresh) {
				onChanged(null, null);
				return;
			}
			try {
				onChanged(null, await fetcher());
			} catch (error) {
				onChanged(error instanceof Error ? error : new Error(String(error)), null);
			}
		};
		const handler = () => {
			if (debounceMs) {
				const existingTimer = this._listChangedDebounceTimers.get(listType);
				if (existingTimer) clearTimeout(existingTimer);
				const timer = setTimeout(refresh, debounceMs);
				this._listChangedDebounceTimers.set(listType, timer);
			} else refresh();
		};
		this.setNotificationHandler(notificationMethod, handler);
	}
	/** Notifies the server that the client's root list has changed. Requires the `roots.listChanged` capability. */
	async sendRootsListChanged() {
		return this.notification({ method: "notifications/roots/list_changed" });
	}
};

//#endregion
//#region src/client/crossAppAccess.ts
/**
* Requests a JWT Authorization Grant (ID-JAG) from an enterprise IdP using RFC 8693 Token Exchange.
*
* This function performs step 2 of the Enterprise Managed Authorization flow:
* exchanges an ID Token for a JWT Authorization Grant that can be used with the target MCP server.
*
* @param options - Configuration for the token exchange request
* @returns The JWT Authorization Grant and related metadata
* @throws {Error} If the token exchange fails or returns an error response
*
* @example
* ```ts
* const result = await requestJwtAuthorizationGrant({
*     tokenEndpoint: 'https://idp.example.com/token',
*     audience: 'https://auth.chat.example/',
*     resource: 'https://mcp.chat.example/',
*     idToken: 'eyJhbGciOiJS...',
*     clientId: 'my-idp-client',
*     clientSecret: 'my-idp-secret',
*     scope: 'chat.read chat.history'
* });
*
* // Use result.jwtAuthGrant with the MCP server's authorization server
* ```
*/
async function requestJwtAuthorizationGrant(options) {
	const { tokenEndpoint, audience, resource, idToken, clientId, clientSecret, scope, fetchFn = fetch } = options;
	const params = new URLSearchParams({
		grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
		requested_token_type: "urn:ietf:params:oauth:token-type:id-jag",
		audience: String(audience),
		resource: String(resource),
		subject_token: idToken,
		subject_token_type: "urn:ietf:params:oauth:token-type:id_token",
		client_id: clientId
	});
	if (clientSecret) params.set("client_secret", clientSecret);
	if (scope) params.set("scope", scope);
	const response = await fetchFn(String(tokenEndpoint), {
		method: "POST",
		headers: { "Content-Type": "application/x-www-form-urlencoded" },
		body: params.toString()
	});
	if (!response.ok) {
		const errorBody = await response.json().catch(() => ({}));
		const parseResult$1 = OAuthErrorResponseSchema.safeParse(errorBody);
		if (parseResult$1.success) {
			const { error, error_description } = parseResult$1.data;
			throw new Error(`Token exchange failed: ${error}${error_description ? ` - ${error_description}` : ""}`);
		}
		throw new Error(`Token exchange failed with status ${response.status}: ${JSON.stringify(errorBody)}`);
	}
	const parseResult = IdJagTokenExchangeResponseSchema.safeParse(await response.json());
	if (!parseResult.success) throw new Error(`Invalid token exchange response: ${parseResult.error.message}`);
	return {
		jwtAuthGrant: parseResult.data.access_token,
		expiresIn: parseResult.data.expires_in,
		scope: parseResult.data.scope
	};
}
/**
* Discovers the IdP's token endpoint and requests a JWT Authorization Grant.
*
* This is a convenience wrapper around {@linkcode requestJwtAuthorizationGrant} that
* first performs OAuth metadata discovery to find the token endpoint.
*
* @param options - Configuration including IdP URL for discovery
* @returns The JWT Authorization Grant and related metadata
* @throws {Error} If discovery fails or the token exchange fails
*
* @example
* ```ts
* const result = await discoverAndRequestJwtAuthGrant({
*     idpUrl: 'https://idp.example.com',
*     audience: 'https://auth.chat.example/',
*     resource: 'https://mcp.chat.example/',
*     idToken: await getIdToken(),
*     clientId: 'my-idp-client',
*     clientSecret: 'my-idp-secret'
* });
* ```
*/
async function discoverAndRequestJwtAuthGrant(options) {
	const { idpUrl, fetchFn = fetch, ...restOptions } = options;
	const metadata = await discoverAuthorizationServerMetadata(String(idpUrl), { fetchFn });
	if (!metadata?.token_endpoint) throw new Error(`Failed to discover token endpoint for IdP: ${idpUrl}`);
	return requestJwtAuthorizationGrant({
		...restOptions,
		tokenEndpoint: metadata.token_endpoint,
		fetchFn
	});
}
/**
* Exchanges a JWT Authorization Grant for an access token at the MCP server's authorization server.
*
* This function performs step 3 of the Enterprise Managed Authorization flow:
* uses the JWT Authorization Grant to obtain an access token from the MCP server.
*
* @param options - Configuration for the JWT grant exchange
* @returns OAuth tokens (access token, token type, etc.)
* @throws {Error} If the exchange fails or returns an error response
*
* Defaults to `client_secret_basic` (HTTP Basic Authorization header), matching
* `CrossAppAccessProvider`'s declared `token_endpoint_auth_method` and the
* SEP-990 conformance test requirements. Use `authMethod: 'client_secret_post'` only
* when the authorization server explicitly requires it.
*
* @example
* ```ts
* const tokens = await exchangeJwtAuthGrant({
*     tokenEndpoint: 'https://auth.chat.example/token',
*     jwtAuthGrant: 'eyJhbGci...',
*     clientId: 'my-mcp-client',
*     clientSecret: 'my-mcp-secret'
* });
*
* // Use tokens.access_token to access the MCP server
* ```
*/
async function exchangeJwtAuthGrant(options) {
	const { tokenEndpoint, jwtAuthGrant, clientId, clientSecret, authMethod = "client_secret_basic", fetchFn = fetch } = options;
	const params = new URLSearchParams({
		grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
		assertion: jwtAuthGrant
	});
	const headers = new Headers({ "Content-Type": "application/x-www-form-urlencoded" });
	applyClientAuthentication(authMethod, {
		client_id: clientId,
		client_secret: clientSecret
	}, headers, params);
	const response = await fetchFn(String(tokenEndpoint), {
		method: "POST",
		headers,
		body: params.toString()
	});
	if (!response.ok) {
		const errorBody = await response.json().catch(() => ({}));
		const parseResult$1 = OAuthErrorResponseSchema.safeParse(errorBody);
		if (parseResult$1.success) {
			const { error, error_description } = parseResult$1.data;
			throw new Error(`JWT grant exchange failed: ${error}${error_description ? ` - ${error_description}` : ""}`);
		}
		throw new Error(`JWT grant exchange failed with status ${response.status}: ${JSON.stringify(errorBody)}`);
	}
	const responseBody = await response.json();
	const parseResult = OAuthTokensSchema.safeParse(responseBody);
	if (!parseResult.success) throw new Error(`Invalid token response: ${parseResult.error.message}`);
	return parseResult.data;
}

//#endregion
//#region src/client/middleware.ts
/**
* Creates a fetch wrapper that handles OAuth authentication automatically.
*
* This wrapper will:
* - Add `Authorization` headers with access tokens
* - Handle 401 responses by attempting re-authentication
* - Retry the original request after successful auth
* - Handle OAuth errors appropriately ({@linkcode index.OAuthErrorCode.InvalidClient | OAuthErrorCode.InvalidClient}, etc.)
*
* The `baseUrl` parameter is optional and defaults to using the domain from the request URL.
* However, you should explicitly provide `baseUrl` when:
* - Making requests to multiple subdomains (e.g., api.example.com, cdn.example.com)
* - Using API paths that differ from OAuth discovery paths (e.g., requesting /api/v1/data but OAuth is at /)
* - The OAuth server is on a different domain than your API requests
* - You want to ensure consistent OAuth behavior regardless of request URLs
*
* For MCP transports, set `baseUrl` to the same URL you pass to the transport constructor.
*
* Note: This wrapper is designed for general-purpose fetch operations.
* MCP transports (SSE and StreamableHTTP) already have built-in OAuth handling
* and should not need this wrapper.
*
* @param provider - OAuth client provider for authentication
* @param baseUrl - Base URL for OAuth server discovery (defaults to request URL domain)
* @returns A fetch middleware function
*/
const withOAuth = (provider, baseUrl) => (next) => {
	return async (input, init) => {
		const makeRequest = async () => {
			const headers = new Headers(init?.headers);
			const tokens = await provider.tokens();
			if (tokens) headers.set("Authorization", `Bearer ${tokens.access_token}`);
			return await next(input, {
				...init,
				headers
			});
		};
		let response = await makeRequest();
		if (response.status === 401) try {
			const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(response);
			const result = await auth(provider, {
				serverUrl: baseUrl || (typeof input === "string" ? new URL(input).origin : input.origin),
				resourceMetadataUrl,
				scope,
				fetchFn: next
			});
			if (result === "REDIRECT") throw new UnauthorizedError("Authentication requires user authorization - redirect initiated");
			if (result !== "AUTHORIZED") throw new UnauthorizedError(`Authentication failed with result: ${result}`);
			response = await makeRequest();
		} catch (error) {
			if (error instanceof UnauthorizedError) throw error;
			throw new UnauthorizedError(`Failed to re-authenticate: ${error instanceof Error ? error.message : String(error)}`);
		}
		if (response.status === 401) throw new UnauthorizedError(`Authentication failed for ${typeof input === "string" ? input : input.toString()}`);
		return response;
	};
};
/**
* Creates a fetch middleware that logs HTTP requests and responses.
*
* When called without arguments `withLogging()`, it uses the default logger that:
* - Logs successful requests (2xx) to `console.log`
* - Logs error responses (4xx/5xx) and network errors to `console.error`
* - Logs all requests regardless of status (`statusLevel: 0`)
* - Does not include request or response headers in logs
* - Measures and displays request duration in milliseconds
*
* Important: the default logger uses both `console.log` and `console.error` so it should not be used with
* `stdio` transports and applications.
*
* @param options - Logging configuration options
* @returns A fetch middleware function
*/
const withLogging = (options = {}) => {
	const { logger, includeRequestHeaders = false, includeResponseHeaders = false, statusLevel = 0 } = options;
	const defaultLogger = (input) => {
		const { method, url, status, statusText, duration, requestHeaders, responseHeaders, error } = input;
		let message = error ? `HTTP ${method} ${url} failed: ${error.message} (${duration}ms)` : `HTTP ${method} ${url} ${status} ${statusText} (${duration}ms)`;
		if (includeRequestHeaders && requestHeaders) {
			const reqHeaders = [...requestHeaders.entries()].map(([key, value]) => `${key}: ${value}`).join(", ");
			message += `\n  Request Headers: {${reqHeaders}}`;
		}
		if (includeResponseHeaders && responseHeaders) {
			const resHeaders = [...responseHeaders.entries()].map(([key, value]) => `${key}: ${value}`).join(", ");
			message += `\n  Response Headers: {${resHeaders}}`;
		}
		if (error || status >= 400) console.error(message);
		else console.log(message);
	};
	const logFn = logger || defaultLogger;
	return (next) => async (input, init) => {
		const startTime = performance.now();
		const method = init?.method || "GET";
		const url = typeof input === "string" ? input : input.toString();
		const requestHeaders = includeRequestHeaders ? new Headers(init?.headers) : void 0;
		try {
			const response = await next(input, init);
			const duration = performance.now() - startTime;
			if (response.status >= statusLevel) logFn({
				method,
				url,
				status: response.status,
				statusText: response.statusText,
				duration,
				requestHeaders,
				responseHeaders: includeResponseHeaders ? response.headers : void 0
			});
			return response;
		} catch (error) {
			logFn({
				method,
				url,
				status: 0,
				statusText: "Network Error",
				duration: performance.now() - startTime,
				requestHeaders,
				error
			});
			throw error;
		}
	};
};
/**
* Composes multiple fetch middleware functions into a single middleware pipeline.
* Middleware are applied in the order they appear, creating a chain of handlers.
*
* @example
* ```ts source="./middleware.examples.ts#applyMiddlewares_basicUsage"
* // Create a middleware pipeline that handles both OAuth and logging
* const enhancedFetch = applyMiddlewares(withOAuth(oauthProvider, 'https://api.example.com'), withLogging({ statusLevel: 400 }))(fetch);
*
* // Use the enhanced fetch - it will handle auth and log errors
* const response = await enhancedFetch('https://api.example.com/data');
* ```
*
* @param middleware - Array of fetch middleware to compose into a pipeline
* @returns A single composed middleware function
*/
const applyMiddlewares = (...middleware) => {
	return (next) => {
		let handler = next;
		for (const mw of middleware) handler = mw(handler);
		return handler;
	};
};
/**
* Helper function to create custom fetch middleware with cleaner syntax.
* Provides the next handler and request details as separate parameters for easier access.
*
* @example
* ```ts source="./middleware.examples.ts#createMiddleware_examples"
* // Create custom authentication middleware
* const customAuthMiddleware = createMiddleware(async (next, input, init) => {
*     const headers = new Headers(init?.headers);
*     headers.set('X-Custom-Auth', 'my-token');
*
*     const response = await next(input, { ...init, headers });
*
*     if (response.status === 401) {
*         console.log('Authentication failed');
*     }
*
*     return response;
* });
*
* // Create conditional middleware
* const conditionalMiddleware = createMiddleware(async (next, input, init) => {
*     const url = typeof input === 'string' ? input : input.toString();
*
*     // Only add headers for API routes
*     if (url.includes('/api/')) {
*         const headers = new Headers(init?.headers);
*         headers.set('X-API-Version', 'v2');
*         return next(input, { ...init, headers });
*     }
*
*     // Pass through for non-API routes
*     return next(input, init);
* });
*
* // Create caching middleware
* const cacheMiddleware = createMiddleware(async (next, input, init) => {
*     const cacheKey = typeof input === 'string' ? input : input.toString();
*
*     // Check cache first
*     const cached = await getFromCache(cacheKey);
*     if (cached) {
*         return new Response(cached, { status: 200 });
*     }
*
*     // Make request and cache result
*     const response = await next(input, init);
*     if (response.ok) {
*         await saveToCache(cacheKey, await response.clone().text());
*     }
*
*     return response;
* });
* ```
*
* @param handler - Function that receives the next handler and request parameters
* @returns A fetch middleware function
*/
const createMiddleware = (handler) => {
	return (next) => (input, init) => handler(next, input, init);
};

//#endregion
//#region src/client/sse.ts
var SseError = class extends Error {
	constructor(code, message, event) {
		super(`SSE error: ${message}`);
		this.code = code;
		this.event = event;
	}
};
/**
* Client transport for SSE: this will connect to a server using Server-Sent Events for receiving
* messages and make separate `POST` requests for sending messages.
* @deprecated SSEClientTransport is deprecated. Prefer to use {@linkcode index.StreamableHTTPClientTransport | StreamableHTTPClientTransport} where possible instead. Note that because some servers are still using SSE, clients may need to support both transports during the migration period.
*/
var SSEClientTransport = class {
	_eventSource;
	_endpoint;
	_abortController;
	_url;
	_resourceMetadataUrl;
	_scope;
	_eventSourceInit;
	_requestInit;
	_authProvider;
	_oauthProvider;
	_fetch;
	_fetchWithInit;
	_protocolVersion;
	onclose;
	onerror;
	onmessage;
	constructor(url, opts) {
		this._url = url;
		this._resourceMetadataUrl = void 0;
		this._scope = void 0;
		this._eventSourceInit = opts?.eventSourceInit;
		this._requestInit = opts?.requestInit;
		if (isOAuthClientProvider(opts?.authProvider)) {
			this._oauthProvider = opts.authProvider;
			this._authProvider = adaptOAuthProvider(opts.authProvider);
		} else this._authProvider = opts?.authProvider;
		this._fetch = opts?.fetch;
		this._fetchWithInit = createFetchWithInit(opts?.fetch, opts?.requestInit);
	}
	_last401Response;
	async _commonHeaders() {
		const headers = {};
		const token = await this._authProvider?.token();
		if (token) headers["Authorization"] = `Bearer ${token}`;
		if (this._protocolVersion) headers["mcp-protocol-version"] = this._protocolVersion;
		const extraHeaders = normalizeHeaders(this._requestInit?.headers);
		return new Headers({
			...headers,
			...extraHeaders
		});
	}
	_startOrAuth() {
		const fetchImpl = this?._eventSourceInit?.fetch ?? this._fetch ?? fetch;
		return new Promise((resolve, reject) => {
			this._eventSource = new EventSource(this._url.href, {
				...this._eventSourceInit,
				fetch: async (url, init) => {
					const headers = await this._commonHeaders();
					headers.set("Accept", "text/event-stream");
					const response = await fetchImpl(url, {
						...init,
						headers
					});
					if (response.status === 401) {
						this._last401Response = response;
						if (response.headers.has("www-authenticate")) {
							const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(response);
							this._resourceMetadataUrl = resourceMetadataUrl;
							this._scope = scope;
						}
					}
					return response;
				}
			});
			this._abortController = new AbortController();
			this._eventSource.onerror = (event) => {
				if (event.code === 401 && this._authProvider) {
					if (this._authProvider.onUnauthorized && this._last401Response) {
						const response = this._last401Response;
						this._last401Response = void 0;
						this._eventSource?.close();
						this._authProvider.onUnauthorized({
							response,
							serverUrl: this._url,
							fetchFn: this._fetchWithInit
						}).then(() => this._startOrAuth().then(resolve, reject), (error$2) => {
							this.onerror?.(error$2);
							reject(error$2);
						});
						return;
					}
					const error$1 = new UnauthorizedError();
					reject(error$1);
					this.onerror?.(error$1);
					return;
				}
				const error = new SseError(event.code, event.message, event);
				reject(error);
				this.onerror?.(error);
			};
			this._eventSource.onopen = () => {};
			this._eventSource.addEventListener("endpoint", (event) => {
				const messageEvent = event;
				try {
					this._endpoint = new URL(messageEvent.data, this._url);
					if (this._endpoint.origin !== this._url.origin) throw new Error(`Endpoint origin does not match connection origin: ${this._endpoint.origin}`);
				} catch (error) {
					reject(error);
					this.onerror?.(error);
					this.close();
					return;
				}
				resolve();
			});
			this._eventSource.onmessage = (event) => {
				const messageEvent = event;
				let message;
				try {
					message = JSONRPCMessageSchema.parse(JSON.parse(messageEvent.data));
				} catch (error) {
					this.onerror?.(error);
					return;
				}
				this.onmessage?.(message);
			};
		});
	}
	async start() {
		if (this._eventSource) throw new Error("SSEClientTransport already started! If using Client class, note that connect() calls start() automatically.");
		return await this._startOrAuth();
	}
	/**
	* Call this method after the user has finished authorizing via their user agent and is redirected back to the MCP client application. This will exchange the authorization code for an access token, enabling the next connection attempt to successfully auth.
	*/
	async finishAuth(authorizationCode) {
		if (!this._oauthProvider) throw new UnauthorizedError("finishAuth requires an OAuthClientProvider");
		if (await auth(this._oauthProvider, {
			serverUrl: this._url,
			authorizationCode,
			resourceMetadataUrl: this._resourceMetadataUrl,
			scope: this._scope,
			fetchFn: this._fetchWithInit
		}) !== "AUTHORIZED") throw new UnauthorizedError("Failed to authorize");
	}
	async close() {
		this._abortController?.abort();
		this._eventSource?.close();
		this.onclose?.();
	}
	async send(message) {
		return this._send(message, false);
	}
	async _send(message, isAuthRetry) {
		if (!this._endpoint) throw new SdkError(SdkErrorCode.NotConnected, "Not connected");
		try {
			const headers = await this._commonHeaders();
			headers.set("content-type", "application/json");
			const init = {
				...this._requestInit,
				method: "POST",
				headers,
				body: JSON.stringify(message),
				signal: this._abortController?.signal
			};
			const response = await (this._fetch ?? fetch)(this._endpoint, init);
			if (!response.ok) {
				if (response.status === 401 && this._authProvider) {
					if (response.headers.has("www-authenticate")) {
						const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(response);
						this._resourceMetadataUrl = resourceMetadataUrl;
						this._scope = scope;
					}
					if (this._authProvider.onUnauthorized && !isAuthRetry) {
						await this._authProvider.onUnauthorized({
							response,
							serverUrl: this._url,
							fetchFn: this._fetchWithInit
						});
						await response.text?.().catch(() => {});
						return this._send(message, true);
					}
					await response.text?.().catch(() => {});
					if (isAuthRetry) throw new SdkError(SdkErrorCode.ClientHttpAuthentication, "Server returned 401 after re-authentication", { status: 401 });
					throw new UnauthorizedError();
				}
				const text = await response.text?.().catch(() => null);
				throw new Error(`Error POSTing to endpoint (HTTP ${response.status}): ${text}`);
			}
			await response.text?.().catch(() => {});
		} catch (error) {
			this.onerror?.(error);
			throw error;
		}
	}
	setProtocolVersion(version) {
		this._protocolVersion = version;
	}
};

//#endregion
//#region src/client/stdio.ts
/**
* Environment variables to inherit by default, if an environment is not explicitly given.
*/
const DEFAULT_INHERITED_ENV_VARS = process.platform === "win32" ? [
	"APPDATA",
	"HOMEDRIVE",
	"HOMEPATH",
	"LOCALAPPDATA",
	"PATH",
	"PROCESSOR_ARCHITECTURE",
	"SYSTEMDRIVE",
	"SYSTEMROOT",
	"TEMP",
	"USERNAME",
	"USERPROFILE",
	"PROGRAMFILES"
] : [
	"HOME",
	"LOGNAME",
	"PATH",
	"SHELL",
	"TERM",
	"USER"
];
/**
* Returns a default environment object including only environment variables deemed safe to inherit.
*/
function getDefaultEnvironment() {
	const env = {};
	for (const key of DEFAULT_INHERITED_ENV_VARS) {
		const value = process.env[key];
		if (value === void 0) continue;
		if (value.startsWith("()")) continue;
		env[key] = value;
	}
	return env;
}
/**
* Client transport for stdio: this will connect to a server by spawning a process and communicating with it over stdin/stdout.
*
* This transport is only available in Node.js environments.
*/
var StdioClientTransport = class {
	_process;
	_readBuffer = new ReadBuffer();
	_serverParams;
	_stderrStream = null;
	onclose;
	onerror;
	onmessage;
	constructor(server) {
		this._serverParams = server;
		if (server.stderr === "pipe" || server.stderr === "overlapped") this._stderrStream = new PassThrough();
	}
	/**
	* Starts the server process and prepares to communicate with it.
	*/
	async start() {
		if (this._process) throw new Error("StdioClientTransport already started! If using Client class, note that connect() calls start() automatically.");
		return new Promise((resolve, reject) => {
			this._process = spawn(this._serverParams.command, this._serverParams.args ?? [], {
				env: {
					...getDefaultEnvironment(),
					...this._serverParams.env
				},
				stdio: [
					"pipe",
					"pipe",
					this._serverParams.stderr ?? "inherit"
				],
				shell: false,
				windowsHide: process.platform === "win32",
				cwd: this._serverParams.cwd
			});
			this._process.on("error", (error) => {
				reject(error);
				this.onerror?.(error);
			});
			this._process.on("spawn", () => {
				resolve();
			});
			this._process.on("close", (_code) => {
				this._process = void 0;
				this.onclose?.();
			});
			this._process.stdin?.on("error", (error) => {
				this.onerror?.(error);
			});
			this._process.stdout?.on("data", (chunk) => {
				this._readBuffer.append(chunk);
				this.processReadBuffer();
			});
			this._process.stdout?.on("error", (error) => {
				this.onerror?.(error);
			});
			if (this._stderrStream && this._process.stderr) this._process.stderr.pipe(this._stderrStream);
		});
	}
	/**
	* The `stderr` stream of the child process, if {@linkcode StdioServerParameters.stderr} was set to `"pipe"` or `"overlapped"`.
	*
	* If `stderr` piping was requested, a `PassThrough` stream is returned _immediately_, allowing callers to
	* attach listeners before the `start` method is invoked. This prevents loss of any early
	* error output emitted by the child process.
	*/
	get stderr() {
		if (this._stderrStream) return this._stderrStream;
		return this._process?.stderr ?? null;
	}
	/**
	* The child process pid spawned by this transport.
	*
	* This is only available after the transport has been started.
	*/
	get pid() {
		return this._process?.pid ?? null;
	}
	processReadBuffer() {
		while (true) try {
			const message = this._readBuffer.readMessage();
			if (message === null) break;
			this.onmessage?.(message);
		} catch (error) {
			this.onerror?.(error);
		}
	}
	async close() {
		if (this._process) {
			const processToClose = this._process;
			this._process = void 0;
			const closePromise = new Promise((resolve) => {
				processToClose.once("close", () => {
					resolve();
				});
			});
			try {
				processToClose.stdin?.end();
			} catch {}
			await Promise.race([closePromise, new Promise((resolve) => setTimeout(resolve, 2e3).unref())]);
			if (processToClose.exitCode === null) {
				try {
					processToClose.kill("SIGTERM");
				} catch {}
				await Promise.race([closePromise, new Promise((resolve) => setTimeout(resolve, 2e3).unref())]);
			}
			if (processToClose.exitCode === null) try {
				processToClose.kill("SIGKILL");
			} catch {}
		}
		this._readBuffer.clear();
	}
	send(message) {
		return new Promise((resolve) => {
			if (!this._process?.stdin) throw new SdkError(SdkErrorCode.NotConnected, "Not connected");
			const json = serializeMessage(message);
			if (this._process.stdin.write(json)) resolve();
			else this._process.stdin.once("drain", resolve);
		});
	}
};

//#endregion
//#region src/client/streamableHttp.ts
const DEFAULT_STREAMABLE_HTTP_RECONNECTION_OPTIONS = {
	initialReconnectionDelay: 1e3,
	maxReconnectionDelay: 3e4,
	reconnectionDelayGrowFactor: 1.5,
	maxRetries: 2
};
/**
* Client transport for Streamable HTTP: this implements the MCP Streamable HTTP transport specification.
* It will connect to a server using HTTP `POST` for sending messages and HTTP `GET` with Server-Sent Events
* for receiving messages.
*/
var StreamableHTTPClientTransport = class {
	_abortController;
	_url;
	_resourceMetadataUrl;
	_scope;
	_requestInit;
	_authProvider;
	_oauthProvider;
	_fetch;
	_fetchWithInit;
	_sessionId;
	_reconnectionOptions;
	_protocolVersion;
	_lastUpscopingHeader;
	_serverRetryMs;
	_reconnectionScheduler;
	_cancelReconnection;
	onclose;
	onerror;
	onmessage;
	constructor(url, opts) {
		this._url = url;
		this._resourceMetadataUrl = void 0;
		this._scope = void 0;
		this._requestInit = opts?.requestInit;
		if (isOAuthClientProvider(opts?.authProvider)) {
			this._oauthProvider = opts.authProvider;
			this._authProvider = adaptOAuthProvider(opts.authProvider);
		} else this._authProvider = opts?.authProvider;
		this._fetch = opts?.fetch;
		this._fetchWithInit = createFetchWithInit(opts?.fetch, opts?.requestInit);
		this._sessionId = opts?.sessionId;
		this._protocolVersion = opts?.protocolVersion;
		this._reconnectionOptions = opts?.reconnectionOptions ?? DEFAULT_STREAMABLE_HTTP_RECONNECTION_OPTIONS;
		this._reconnectionScheduler = opts?.reconnectionScheduler;
	}
	async _commonHeaders() {
		const headers = {};
		const token = await this._authProvider?.token();
		if (token) headers["Authorization"] = `Bearer ${token}`;
		if (this._sessionId) headers["mcp-session-id"] = this._sessionId;
		if (this._protocolVersion) headers["mcp-protocol-version"] = this._protocolVersion;
		const extraHeaders = normalizeHeaders(this._requestInit?.headers);
		return new Headers({
			...headers,
			...extraHeaders
		});
	}
	async _startOrAuthSse(options, isAuthRetry = false) {
		const { resumptionToken } = options;
		try {
			const headers = await this._commonHeaders();
			headers.set("Accept", "text/event-stream");
			if (resumptionToken) headers.set("last-event-id", resumptionToken);
			const response = await (this._fetch ?? fetch)(this._url, {
				...this._requestInit,
				method: "GET",
				headers,
				signal: this._abortController?.signal
			});
			if (!response.ok) {
				if (response.status === 401 && this._authProvider) {
					if (response.headers.has("www-authenticate")) {
						const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(response);
						this._resourceMetadataUrl = resourceMetadataUrl;
						this._scope = scope;
					}
					if (this._authProvider.onUnauthorized && !isAuthRetry) {
						await this._authProvider.onUnauthorized({
							response,
							serverUrl: this._url,
							fetchFn: this._fetchWithInit
						});
						await response.text?.().catch(() => {});
						return this._startOrAuthSse(options, true);
					}
					await response.text?.().catch(() => {});
					if (isAuthRetry) throw new SdkError(SdkErrorCode.ClientHttpAuthentication, "Server returned 401 after re-authentication", { status: 401 });
					throw new UnauthorizedError();
				}
				await response.text?.().catch(() => {});
				if (response.status === 405) return;
				throw new SdkError(SdkErrorCode.ClientHttpFailedToOpenStream, `Failed to open SSE stream: ${response.statusText}`, {
					status: response.status,
					statusText: response.statusText
				});
			}
			this._handleSseStream(response.body, options, true);
		} catch (error) {
			this.onerror?.(error);
			throw error;
		}
	}
	/**
	* Calculates the next reconnection delay using a backoff algorithm
	*
	* @param attempt Current reconnection attempt count for the specific stream
	* @returns Time to wait in milliseconds before next reconnection attempt
	*/
	_getNextReconnectionDelay(attempt) {
		if (this._serverRetryMs !== void 0) return this._serverRetryMs;
		const initialDelay = this._reconnectionOptions.initialReconnectionDelay;
		const growFactor = this._reconnectionOptions.reconnectionDelayGrowFactor;
		const maxDelay = this._reconnectionOptions.maxReconnectionDelay;
		return Math.min(initialDelay * Math.pow(growFactor, attempt), maxDelay);
	}
	/**
	* Schedule a reconnection attempt using server-provided retry interval or backoff
	*
	* @param lastEventId The ID of the last received event for resumability
	* @param attemptCount Current reconnection attempt count for this specific stream
	*/
	_scheduleReconnection(options, attemptCount = 0) {
		const maxRetries = this._reconnectionOptions.maxRetries;
		if (attemptCount >= maxRetries) {
			this.onerror?.(/* @__PURE__ */ new Error(`Maximum reconnection attempts (${maxRetries}) exceeded.`));
			return;
		}
		const delay = this._getNextReconnectionDelay(attemptCount);
		const reconnect = () => {
			this._cancelReconnection = void 0;
			if (this._abortController?.signal.aborted) return;
			this._startOrAuthSse(options).catch((error) => {
				this.onerror?.(/* @__PURE__ */ new Error(`Failed to reconnect SSE stream: ${error instanceof Error ? error.message : String(error)}`));
				try {
					this._scheduleReconnection(options, attemptCount + 1);
				} catch (scheduleError) {
					this.onerror?.(scheduleError instanceof Error ? scheduleError : new Error(String(scheduleError)));
				}
			});
		};
		if (this._reconnectionScheduler) {
			const cancel = this._reconnectionScheduler(reconnect, delay, attemptCount);
			this._cancelReconnection = typeof cancel === "function" ? cancel : void 0;
		} else {
			const handle = setTimeout(reconnect, delay);
			this._cancelReconnection = () => clearTimeout(handle);
		}
	}
	_handleSseStream(stream, options, isReconnectable) {
		if (!stream) return;
		const { onresumptiontoken, replayMessageId } = options;
		let lastEventId;
		let hasPrimingEvent = false;
		let receivedResponse = false;
		const processStream = async () => {
			try {
				const reader = stream.pipeThrough(new TextDecoderStream()).pipeThrough(new EventSourceParserStream({ onRetry: (retryMs) => {
					this._serverRetryMs = retryMs;
				} })).getReader();
				while (true) {
					const { value: event, done } = await reader.read();
					if (done) break;
					if (event.id) {
						lastEventId = event.id;
						hasPrimingEvent = true;
						onresumptiontoken?.(event.id);
					}
					if (!event.data) continue;
					if (!event.event || event.event === "message") try {
						const message = JSONRPCMessageSchema.parse(JSON.parse(event.data));
						if (isJSONRPCResultResponse(message) || isJSONRPCErrorResponse(message)) {
							receivedResponse = true;
							if (replayMessageId !== void 0) message.id = replayMessageId;
						}
						this.onmessage?.(message);
					} catch (error) {
						this.onerror?.(error);
					}
				}
				if ((isReconnectable || hasPrimingEvent) && !receivedResponse && this._abortController && !this._abortController.signal.aborted) this._scheduleReconnection({
					resumptionToken: lastEventId,
					onresumptiontoken,
					replayMessageId
				}, 0);
			} catch (error) {
				this.onerror?.(/* @__PURE__ */ new Error(`SSE stream disconnected: ${error}`));
				if ((isReconnectable || hasPrimingEvent) && !receivedResponse && this._abortController && !this._abortController.signal.aborted) try {
					this._scheduleReconnection({
						resumptionToken: lastEventId,
						onresumptiontoken,
						replayMessageId
					}, 0);
				} catch (error$1) {
					this.onerror?.(/* @__PURE__ */ new Error(`Failed to reconnect: ${error$1 instanceof Error ? error$1.message : String(error$1)}`));
				}
			}
		};
		processStream();
	}
	async start() {
		if (this._abortController) throw new Error("StreamableHTTPClientTransport already started! If using Client class, note that connect() calls start() automatically.");
		this._abortController = new AbortController();
	}
	/**
	* Call this method after the user has finished authorizing via their user agent and is redirected back to the MCP client application. This will exchange the authorization code for an access token, enabling the next connection attempt to successfully auth.
	*/
	async finishAuth(authorizationCode) {
		if (!this._oauthProvider) throw new UnauthorizedError("finishAuth requires an OAuthClientProvider");
		if (await auth(this._oauthProvider, {
			serverUrl: this._url,
			authorizationCode,
			resourceMetadataUrl: this._resourceMetadataUrl,
			scope: this._scope,
			fetchFn: this._fetchWithInit
		}) !== "AUTHORIZED") throw new UnauthorizedError("Failed to authorize");
	}
	async close() {
		try {
			this._cancelReconnection?.();
		} finally {
			this._cancelReconnection = void 0;
			this._abortController?.abort();
			this.onclose?.();
		}
	}
	async send(message, options) {
		return this._send(message, options, false);
	}
	async _send(message, options, isAuthRetry) {
		try {
			const { resumptionToken, onresumptiontoken } = options || {};
			if (resumptionToken) {
				this._startOrAuthSse({
					resumptionToken,
					replayMessageId: isJSONRPCRequest(message) ? message.id : void 0
				}).catch((error) => this.onerror?.(error));
				return;
			}
			const headers = await this._commonHeaders();
			headers.set("content-type", "application/json");
			headers.set("accept", "application/json, text/event-stream");
			const init = {
				...this._requestInit,
				method: "POST",
				headers,
				body: JSON.stringify(message),
				signal: this._abortController?.signal
			};
			const response = await (this._fetch ?? fetch)(this._url, init);
			const sessionId = response.headers.get("mcp-session-id");
			if (sessionId) this._sessionId = sessionId;
			if (!response.ok) {
				if (response.status === 401 && this._authProvider) {
					if (response.headers.has("www-authenticate")) {
						const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(response);
						this._resourceMetadataUrl = resourceMetadataUrl;
						this._scope = scope;
					}
					if (this._authProvider.onUnauthorized && !isAuthRetry) {
						await this._authProvider.onUnauthorized({
							response,
							serverUrl: this._url,
							fetchFn: this._fetchWithInit
						});
						await response.text?.().catch(() => {});
						return this._send(message, options, true);
					}
					await response.text?.().catch(() => {});
					if (isAuthRetry) throw new SdkError(SdkErrorCode.ClientHttpAuthentication, "Server returned 401 after re-authentication", { status: 401 });
					throw new UnauthorizedError();
				}
				const text = await response.text?.().catch(() => null);
				if (response.status === 403 && this._oauthProvider) {
					const { resourceMetadataUrl, scope, error } = extractWWWAuthenticateParams(response);
					if (error === "insufficient_scope") {
						const wwwAuthHeader = response.headers.get("WWW-Authenticate");
						if (this._lastUpscopingHeader === wwwAuthHeader) throw new SdkError(SdkErrorCode.ClientHttpForbidden, "Server returned 403 after trying upscoping", {
							status: 403,
							text
						});
						if (scope) this._scope = scope;
						if (resourceMetadataUrl) this._resourceMetadataUrl = resourceMetadataUrl;
						this._lastUpscopingHeader = wwwAuthHeader ?? void 0;
						if (await auth(this._oauthProvider, {
							serverUrl: this._url,
							resourceMetadataUrl: this._resourceMetadataUrl,
							scope: this._scope,
							fetchFn: this._fetchWithInit
						}) !== "AUTHORIZED") throw new UnauthorizedError();
						return this._send(message, options, isAuthRetry);
					}
				}
				throw new SdkError(SdkErrorCode.ClientHttpNotImplemented, `Error POSTing to endpoint: ${text}`, {
					status: response.status,
					text
				});
			}
			this._lastUpscopingHeader = void 0;
			if (response.status === 202) {
				await response.text?.().catch(() => {});
				if (isInitializedNotification(message)) this._startOrAuthSse({ resumptionToken: void 0 }).catch((error) => this.onerror?.(error));
				return;
			}
			const hasRequests = (Array.isArray(message) ? message : [message]).some((msg) => "method" in msg && "id" in msg && msg.id !== void 0);
			const contentType = response.headers.get("content-type");
			if (hasRequests) if (contentType?.includes("text/event-stream")) this._handleSseStream(response.body, { onresumptiontoken }, false);
			else if (contentType?.includes("application/json")) {
				const data = await response.json();
				const responseMessages = Array.isArray(data) ? data.map((msg) => JSONRPCMessageSchema.parse(msg)) : [JSONRPCMessageSchema.parse(data)];
				for (const msg of responseMessages) this.onmessage?.(msg);
			} else {
				await response.text?.().catch(() => {});
				throw new SdkError(SdkErrorCode.ClientHttpUnexpectedContent, `Unexpected content type: ${contentType}`, { contentType });
			}
			else await response.text?.().catch(() => {});
		} catch (error) {
			this.onerror?.(error);
			throw error;
		}
	}
	get sessionId() {
		return this._sessionId;
	}
	/**
	* Terminates the current session by sending a `DELETE` request to the server.
	*
	* Clients that no longer need a particular session
	* (e.g., because the user is leaving the client application) SHOULD send an
	* HTTP `DELETE` to the MCP endpoint with the `Mcp-Session-Id` header to explicitly
	* terminate the session.
	*
	* The server MAY respond with HTTP `405 Method Not Allowed`, indicating that
	* the server does not allow clients to terminate sessions.
	*/
	async terminateSession() {
		if (!this._sessionId) return;
		try {
			const headers = await this._commonHeaders();
			const init = {
				...this._requestInit,
				method: "DELETE",
				headers,
				signal: this._abortController?.signal
			};
			const response = await (this._fetch ?? fetch)(this._url, init);
			await response.text?.().catch(() => {});
			if (!response.ok && response.status !== 405) throw new SdkError(SdkErrorCode.ClientHttpFailedToTerminateSession, `Failed to terminate session: ${response.statusText}`, {
				status: response.status,
				statusText: response.statusText
			});
			this._sessionId = void 0;
		} catch (error) {
			this.onerror?.(error);
			throw error;
		}
	}
	setProtocolVersion(version) {
		this._protocolVersion = version;
	}
	get protocolVersion() {
		return this._protocolVersion;
	}
	/**
	* Resume an SSE stream from a previous event ID.
	* Opens a `GET` SSE connection with `Last-Event-ID` header to replay missed events.
	*
	* @param lastEventId The event ID to resume from
	* @param options Optional callback to receive new resumption tokens
	*/
	async resumeStream(lastEventId, options) {
		await this._startOrAuthSse({
			resumptionToken: lastEventId,
			onresumptiontoken: options?.onresumptiontoken
		});
	}
};

//#endregion
//#region src/fromJsonSchema.ts
let _defaultValidator;
function fromJsonSchema(schema, validator) {
	return fromJsonSchema$1(schema, validator ?? (_defaultValidator ??= new DefaultJsonSchemaValidator()));
}

//#endregion
export { AjvJsonSchemaValidator, CfWorkerJsonSchemaValidator, Client, ClientCredentialsProvider, CrossAppAccessProvider, DEFAULT_INHERITED_ENV_VARS, DEFAULT_NEGOTIATED_PROTOCOL_VERSION, DEFAULT_REQUEST_TIMEOUT_MSEC, ExperimentalClientTasks, INTERNAL_ERROR, INVALID_PARAMS, INVALID_REQUEST, InMemoryTaskMessageQueue, InMemoryTaskStore, JSONRPC_VERSION, LATEST_PROTOCOL_VERSION, METHOD_NOT_FOUND, OAuthError, OAuthErrorCode, PARSE_ERROR, PrivateKeyJwtProvider, ProtocolError, ProtocolErrorCode, RELATED_TASK_META_KEY, ReadBuffer, SSEClientTransport, SUPPORTED_PROTOCOL_VERSIONS, SdkError, SdkErrorCode, SseError, StaticPrivateKeyJwtProvider, StdioClientTransport, StreamableHTTPClientTransport, UnauthorizedError, UriTemplate, UrlElicitationRequiredError, applyMiddlewares, assertClientRequestTaskCapability, assertCompleteRequestPrompt, assertCompleteRequestResourceTemplate, assertToolsCallTaskCapability, auth, buildDiscoveryUrls, checkResourceAllowed, createFetchWithInit, createMiddleware, createPrivateKeyJwtAuth, deserializeMessage, discoverAndRequestJwtAuthGrant, discoverAuthorizationServerMetadata, discoverOAuthMetadata, discoverOAuthProtectedResourceMetadata, discoverOAuthServerInfo, exchangeAuthorization, exchangeJwtAuthGrant, extractResourceMetadataUrl, extractWWWAuthenticateParams, fetchToken, fromJsonSchema, getDefaultEnvironment, getDisplayName, getSupportedElicitationModes, isHttpsUrl, isInitializeRequest, isInitializedNotification, isJSONRPCErrorResponse, isJSONRPCNotification, isJSONRPCRequest, isJSONRPCResultResponse, isTaskAugmentedRequestParams, isTerminal, parseErrorResponse, parseJSONRPCMessage, prepareAuthorizationCodeRequest, refreshAuthorization, registerClient, requestJwtAuthorizationGrant, resourceUrlFromServerUrl, selectClientAuthMethod, selectResourceURL, serializeMessage, startAuthorization, takeResult, toArrayAsync, withLogging, withOAuth };
